{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plots for mutation growth rate paper\n",
    "\n",
    "This notebook generates plots for the [paper/](paper/) directory. This assumes you've alread run\n",
    "```sh\n",
    "make update                       # Downloads and preprocesses data.\n",
    "python mutrans.py --vary-holdout  # Fits and crossvalidates model.\n",
    "```\n",
    "Note that `make update` takes a couple hours the first time it is run (mostly in sequence alignment), and `mutrans.py` takes about 15 minutes on a GPU (will take much longer if no GPU is available)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import math\n",
    "import os\n",
    "import pickle\n",
    "import re\n",
    "import logging\n",
    "from collections import Counter, OrderedDict, defaultdict\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import torch\n",
    "import pyro.distributions as dist\n",
    "from pyrocov import mutrans, pangolin, stats\n",
    "from pyrocov.stats import normal_log10bf\n",
    "from pyrocov.util import pretty_print, pearson_correlation\n",
    "\n",
    "logging.basicConfig(format=\"%(relativeCreated) 9d %(message)s\", level=logging.INFO)\n",
    "matplotlib.rcParams[\"figure.dpi\"] = 200\n",
    "matplotlib.rcParams[\"axes.edgecolor\"] = \"gray\"\n",
    "matplotlib.rcParams[\"savefig.bbox\"] = \"tight\"\n",
    "matplotlib.rcParams['font.family'] = 'sans-serif'\n",
    "matplotlib.rcParams['font.sans-serif'] = ['Arial', 'Avenir', 'DejaVu Sans']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "     4327 Loaded 531 x 3621 daily case data, totaling 217872032.0\n",
      "     4328 Joining GISAID and JHU region codes\n",
      "     4697 Matching 1070 GISAID regions to 3621 JHU fuzzy regions\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "daily_cases \tTensor of shape (531, 1070)\n",
      "features \tTensor of shape (1281, 2337)\n",
      "lineage_id \tdict of size 1281\n",
      "lineage_id_inv \tlist of size 1281\n",
      "local_time \tTensor of shape (42, 1070)\n",
      "location_id \tOrderedDict of size 1070\n",
      "mutations \tlist of size 2337\n",
      "weekly_cases \tTensor of shape (42, 1070)\n",
      "weekly_strains \tTensor of shape (42, 1070, 1281)\n",
      "CPU times: user 958 ms, sys: 308 ms, total: 1.27 s\n",
      "Wall time: 1.18 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "def load_data():\n",
    "    filename = \"results/mutrans.data.single.pt\"\n",
    "    if os.path.exists(filename):\n",
    "        dataset = torch.load(filename)\n",
    "    else:\n",
    "        dataset = mutrans.load_gisaid_data()\n",
    "        torch.save(dataset, filename)\n",
    "    dataset.update(mutrans.load_jhu_data(dataset))\n",
    "    return dataset\n",
    "dataset = load_data()\n",
    "locals().update(dataset)\n",
    "for k, v in sorted(dataset.items()):\n",
    "    if isinstance(v, torch.Tensor):\n",
    "        print(f\"{k} \\t{type(v).__name__} of shape {tuple(v.shape)}\")\n",
    "    else:\n",
    "        print(f\"{k} \\t{type(v).__name__} of size {len(v)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42 x 1070 x 1281 = 57568140\n",
      "2160748\n"
     ]
    }
   ],
   "source": [
    "print(\"{} x {} x {} = {}\".format(*weekly_strains.shape, weekly_strains.shape.numel()))\n",
    "print(int(weekly_strains.sum()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded data from 2161248 samples\n"
     ]
    }
   ],
   "source": [
    "with open(\"results/gisaid.columns.pkl\", \"rb\") as f:\n",
    "    columns = pickle.load(f)\n",
    "print(\"Loaded data from {} samples\".format(len(columns[\"lineage\"])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sanity checking case count time series:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if False:\n",
    "    plt.figure(figsize=(8, 3))\n",
    "    plt.plot(weekly_cases, lw=1, alpha=0.5)\n",
    "    plt.yscale(\"symlog\", linthresh=10)\n",
    "    plt.ylim(0, None)\n",
    "    plt.xlim(0, len(weekly_cases) - 1)\n",
    "    plt.xlabel(\"week after 2019-12-01\")\n",
    "    plt.ylabel(\"confirmed cases\");\n",
    "\n",
    "    plt.figure(figsize=(8, 3))\n",
    "    plt.plot(weekly_strains.sum(-1), lw=1, alpha=0.5)\n",
    "    plt.yscale(\"symlog\", linthresh=10)\n",
    "    plt.ylim(0, None)\n",
    "    plt.xlim(0, len(weekly_cases) - 1)\n",
    "    plt.xlabel(\"week after 2019-12-01\")\n",
    "    plt.ylabel(\"sequenced samples\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52 4 1014\n"
     ]
    }
   ],
   "source": [
    "locations = set(location_id)\n",
    "N_usa = sum(1 for k in locations if \"/ USA /\" in k)\n",
    "N_uk = sum(1 for k in locations if \"/ United Kingdom /\" in k)\n",
    "N_other = len(locations) - N_usa - N_uk\n",
    "print(N_usa, N_uk, N_other)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll account for epidemiological dynamics in the form of random drift on top of our logistic growth model. Since random drift is inversely proportional to the local number of infections, we'll need a new data source for the number of infections in each region. We'll use JHU's confirmed case counts time series as a proxy for the number of total infections in each region."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load trained models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "fits = torch.load(\"results/mutrans.pt\", map_location=\"cpu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, ())\n",
      "('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, (('include', (('location', '^Europe'),)),))\n",
      "('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, (('exclude', (('location', '^Europe'),)),))\n"
     ]
    }
   ],
   "source": [
    "for key in fits:\n",
    "    print(key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ": dict of length 9\n",
      "['median']: dict of length 12\n",
      "['median']['coef_scale']: Tensor of shape ()\n",
      "['median']['rate_scale']: Tensor of shape ()\n",
      "['median']['init_loc_scale']: Tensor of shape ()\n",
      "['median']['init_scale']: Tensor of shape ()\n",
      "['median']['coef_decentered']: Tensor of shape (2337,)\n",
      "['median']['coef']: Tensor of shape (2337,)\n",
      "['median']['rate_loc']: Tensor of shape (1281,)\n",
      "['median']['init_loc_decentered']: Tensor of shape (1281,)\n",
      "['median']['init_loc']: Tensor of shape (1281,)\n",
      "['median']['rate']: Tensor of shape (1070, 1281)\n",
      "['median']['init']: Tensor of shape (1070, 1281)\n",
      "['median']['probs']: Tensor of shape (48, 1070, 1281)\n",
      "['mean']: dict of length 12\n",
      "['mean']['init_loc_scale']: Tensor of shape ()\n",
      "['mean']['coef_decentered']: Tensor of shape (2337,)\n",
      "['mean']['coef']: Tensor of shape (2337,)\n",
      "['mean']['probs']: Tensor of shape (48, 1070, 1281)\n",
      "['mean']['rate']: Tensor of shape (1070, 1281)\n",
      "['mean']['init_loc_decentered']: Tensor of shape (1281,)\n",
      "['mean']['rate_scale']: Tensor of shape ()\n",
      "['mean']['init']: Tensor of shape (1070, 1281)\n",
      "['mean']['init_scale']: Tensor of shape ()\n",
      "['mean']['rate_loc']: Tensor of shape (1281,)\n",
      "['mean']['init_loc']: Tensor of shape (1281,)\n",
      "['mean']['coef_scale']: Tensor of shape ()\n",
      "['std']: dict of length 12\n",
      "['std']['init_loc_scale']: Tensor of shape ()\n",
      "['std']['coef_decentered']: Tensor of shape (2337,)\n",
      "['std']['coef']: Tensor of shape (2337,)\n",
      "['std']['probs']: Tensor of shape (48, 1070, 1281)\n",
      "['std']['rate']: Tensor of shape (1070, 1281)\n",
      "['std']['init_loc_decentered']: Tensor of shape (1281,)\n",
      "['std']['rate_scale']: Tensor of shape ()\n",
      "['std']['init']: Tensor of shape (1070, 1281)\n",
      "['std']['init_scale']: Tensor of shape ()\n",
      "['std']['rate_loc']: Tensor of shape (1281,)\n",
      "['std']['init_loc']: Tensor of shape (1281,)\n",
      "['std']['coef_scale']: Tensor of shape ()\n",
      "['losses']: list of length 10001\n",
      "['series']: dict of length 24\n",
      "['series']['local_time']: list of length 10001\n",
      "['series']['coef_centered']: list of length 10001\n",
      "['series']['init_loc_centered']: list of length 10001\n",
      "['series']['rate_centered']: list of length 10001\n",
      "['series']['init_centered']: list of length 10001\n",
      "['series']['Guide.0.loc']: list of length 10001\n",
      "['series']['Guide.0.scale']: list of length 10001\n",
      "['series']['Guide.0.cov_factor']: list of length 10001\n",
      "['series']['Guide.1.locs.rate_scale']: list of length 10001\n",
      "['series']['Guide.1.scales.rate_scale']: list of length 10001\n",
      "['series']['Guide.1.locs.init_loc_scale']: list of length 10001\n",
      "['series']['Guide.1.scales.init_loc_scale']: list of length 10001\n",
      "['series']['Guide.1.locs.init_scale']: list of length 10001\n",
      "['series']['Guide.1.scales.init_scale']: list of length 10001\n",
      "['series']['Guide.1.locs.init_loc_decentered']: list of length 10001\n",
      "['series']['Guide.1.scales.init_loc_decentered']: list of length 10001\n",
      "['series']['Guide.1.locs.rate_decentered']: list of length 10001\n",
      "['series']['Guide.1.scales.rate_decentered']: list of length 10001\n",
      "['series']['Guide.1.locs.init_decentered']: list of length 10001\n",
      "['series']['Guide.1.scales.init_decentered']: list of length 10001\n",
      "['series']['rate_scale']: list of length 10001\n",
      "['series']['init_loc_scale']: list of length 10001\n",
      "['series']['init_scale']: list of length 10001\n",
      "['series']['loss']: list of length 10001\n",
      "['params']: dict of length 20\n",
      "['params']['local_time']: Tensor of shape (1070, 1281)\n",
      "['params']['coef_centered']: Tensor of shape (2337,)\n",
      "['params']['init_loc_centered']: Tensor of shape (1281,)\n",
      "['params']['rate_centered']: Tensor of shape (1281,)\n",
      "['params']['init_centered']: Tensor of shape (1281,)\n",
      "['params']['Guide.0.loc']: Tensor of shape (2337,)\n",
      "['params']['Guide.0.scale']: Tensor of shape (2337,)\n",
      "['params']['Guide.0.cov_factor']: Tensor of shape (2337, 200)\n",
      "['params']['Guide.1.locs.rate_scale']: Tensor of shape ()\n",
      "['params']['Guide.1.scales.rate_scale']: Tensor of shape ()\n",
      "['params']['Guide.1.locs.init_loc_scale']: Tensor of shape ()\n",
      "['params']['Guide.1.scales.init_loc_scale']: Tensor of shape ()\n",
      "['params']['Guide.1.locs.init_scale']: Tensor of shape ()\n",
      "['params']['Guide.1.scales.init_scale']: Tensor of shape ()\n",
      "['params']['Guide.1.locs.init_loc_decentered']: Tensor of shape (1281,)\n",
      "['params']['Guide.1.scales.init_loc_decentered']: Tensor of shape (1281,)\n",
      "['params']['Guide.1.locs.rate_decentered']: Tensor of shape (1070, 1281)\n",
      "['params']['Guide.1.scales.rate_decentered']: Tensor of shape (1070, 1281)\n",
      "['params']['Guide.1.locs.init_decentered']: Tensor of shape (1070, 1281)\n",
      "['params']['Guide.1.scales.init_decentered']: Tensor of shape (1070, 1281)\n",
      "['walltime'] = 562.4560180319986\n",
      "['args']: Namespace\n",
      "['mutations']: list of length 2337\n"
     ]
    }
   ],
   "source": [
    "best_fit = list(fits.values())[0]\n",
    "pretty_print(best_fit, max_items=40)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scale `coef` by 1/100 in all results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, ()).median.coef\n",
      ".('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, ()).mean.coef\n",
      ".('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, ()).std.coef\n",
      ".('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, (('include', (('location', '^Europe'),)),)).median.coef\n",
      ".('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, (('include', (('location', '^Europe'),)),)).mean.coef\n",
      ".('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, (('include', (('location', '^Europe'),)),)).std.coef\n",
      ".('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, (('exclude', (('location', '^Europe'),)),)).median.coef\n",
      ".('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, (('exclude', (('location', '^Europe'),)),)).mean.coef\n",
      ".('reparam-biased', 'coef_scale=0.5', 'custom', 10001, 0.05, 0.1, 10.0, 200, 6, (('exclude', (('location', '^Europe'),)),)).std.coef\n"
     ]
    }
   ],
   "source": [
    "ALREADY_SCALED = set()\n",
    "\n",
    "def scale_tensors(x, names={\"coef\"}, scale=0.01, prefix=\"\"):\n",
    "    if id(x) in ALREADY_SCALED:\n",
    "        return\n",
    "    if isinstance(x, dict):\n",
    "        for k, v in list(x.items()):\n",
    "            if k in names:\n",
    "                print(f\"{prefix}.{k}\")\n",
    "                x[k] = v * scale\n",
    "            elif k == \"diagnostics\":\n",
    "                continue\n",
    "            else:\n",
    "                scale_tensors(v, names, scale, f\"{prefix}.{k}\")\n",
    "    ALREADY_SCALED.add(id(x))\n",
    "                \n",
    "scale_tensors(fits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABBMAAAGuCAYAAADCqk5RAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAB7CAAAewgFu0HU+AABHWElEQVR4nO3deVzU1eL/8fcAjmwK7qEi7ktqZW4tmpr6MEWzTLuWmmap9bWszMpsQbuVeq9Wlrappd665nXJjcwWl9ab5lIqam4IuIKKIC4gnN8f/OZzQYblowygvJ6PB49ozjLnM8fPMPOeM+fjMMYYAQAAAAAAFJBXcQ8AAAAAAABcXQgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALb4FPcASpuLFy/q2LFjkqSAgAB5eZHnAAAAAAA8JyMjQykpKZKkatWqycfnyqMAwoQiduzYMc2aNau4hwEAAAAAKIUeffRR1ahR44r74WNxAAAAAABgCysTilhAQID1+6OPPqpy5coV42gAAAAAANe65ORka4V81vekV4IwoYhl3SOhXLlyKl++fDGOBgAAAABQmhTWvn18zQEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsIUwAQAAAAAA2EKYAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAW3yKewAAAJRUtcdGXnEf0ZPCC2EkAAAAJQsrEwAAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsIUwAQAAAAAA2EKYAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADY4lPcAwAA4FpWe2zkFfcRPSm8EEYCAABQeFiZAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYItPcQ8AAABPqD02sriHAAAAcM1iZQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsIUwAQAAAAAA2EKYAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADY4vEwYdu2bRo0aJBq1qwpp9OpkJAQ9erVS19//fUV9RsdHa3HH39cdevWVdmyZVWlShV17txZ8+fPv6z+vvnmG3l5ecnhcCg6OvqKxgYAAAAAwLXMx5OdL1++XH379lVaWpp129GjR7Vy5UqtXLlSo0aN0rRp02z3u2HDBnXp0kXJycnWbQkJCVqzZo3WrFmjRYsWacGCBfLxKdjhnTx5Ug8//LCMMbbHAgAAAABAaeOxlQlbtmxR//79lZaWptatW2vdunVKSEjQxo0bdc8990iS3n33Xc2YMcNWv3FxcQoPD1dycrIaNGiglStXKj4+Xtu3b9ewYcMkSUuWLNGLL75Y4D5HjBihw4cP2xoHAAAAAACllcfChFdeeUXnzp1T/fr1tWbNGnXo0EGVKlVSq1attGTJEvXr10+SFBERkW2FQX4mTZqkhIQEBQcHa/369QoPD1flypXVtGlTffzxxxozZowkadq0aQX6usK8efO0aNGiyzpGAAAAAABKI4+ECbt27VJkZKQkady4cQoMDMxW7nA4NHXqVHl5eenEiRNasmRJgfpNTEzU7NmzJUmjRo1SSEhIjjoREREKDg5WWlqa5s6dm2d/Bw8e1JNPPimHw6HBgwcXaAwAAAAAAJR2HgkTVq1aJSkzNOjVq5fbOqGhoWrRooUkaenSpQXqd+3atTp//rwkqXfv3m7rBAYGqnPnzvn2m5GRoYceekhJSUl66qmn1LFjxwKNAQAAAACA0s4jYcLWrVslSWFhYapcuXKu9VxhwqZNm2z16+PjoxtvvDHffrdt26bU1FS3daZMmaIffvhBTZo00cSJEwt0/wAAAAAAwENXc3DtVVCnTp0864WFhUnK3FTx4sWL+V59wdVvaGiovL298+03PT1dsbGxqlevXrbyP/74Q6+88op8fHz0r3/9S76+vnnerx1xcXF5lqekpBTafQEAAAAAUBw8EiYkJCRIkipUqJBnvaCgIEmSMUaJiYl5rmK4nH4l6dSpU9nKLly4oAEDBig1NVXjx49Xy5Yt8+zLrtDQ0DzLy5cvr9GjRxfqfQIAAAAAUJQ88jUH174Gfn5+edbLWu5q4+l+X3zxRe3YsUOtW7fWSy+9lO99AgAAAACA7DyyMiGvryAUZ79r1qzRO++8Iz8/P82bNy/fr1VcjtjY2DzLU1JS9MUXXxT6/QIAAAAAUFQ8EiYEBARIyn+1wblz56zf81ttcKX9JiYmavDgwTLGaOLEiWrcuHG+93c5atasmWd5UlKSR+4XAAAAAICi4pEwITg4WJJ0+vTpPOslJiZKylxxkN8+CJfTryRrH4bHH39ccXFx6tSpk0aNGpXvfQEAUFLUHht5xX1ETwovhJEAAABk8sieCQ0bNpQkHTx4MM96MTExkqQaNWrIyyv/obj6jY2NlTEm3359fHwUEhIiSdZXC9auXSsvLy85HI5sPw8//LDVvk6dOnI4HOrYsWO+YwIAAAAAoLTxSJjQvHlzSdL+/fvzXNa/efNmSVKLFi1s9XvhwgVFRUXl22/Tpk3ldDoL1DcAAAAAACgYj4QJPXr0kCSlp6crMtL90szY2Fht3bpVknTXXXcVqN+OHTvK399fkrR8+XK3dVJSUrRmzZoc/SYnJ+f58+GHH1p1d+zYoeTkZK1atapA4wIAAAAAoDTxSJhQt25dtWvXTpIUERGRY48DY4yeffZZZWRkqHLlyho0aFCB+g0MDFSfPn0kSVOnTrW+zpDV+PHjlZiYKKfTqSeeeCJb27x+ypYta9X19/dXYGBggTaFBAAAAACgtPFImCBJb7/9try8vLRnzx61b99e33zzjRISErR582bdd999WrhwoaTMN/+uqzS4NG7cWI0bN9ZDDz2Uo98333xTAQEBOnHihNq3b6/FixcrPj5eO3fu1IgRIzRlyhRJ0qhRo/K9sgIAAAAAALDPI1dzkKRWrVpp9uzZGjZsmLZt26Zu3brlqDN69GiNHDkyx+27d++WJF133XU5ykJDQ7V48WL16dNHMTEx6tu3b446/fr10+TJkwvhKAAAAAAAwKU8tjJBkoYMGaLNmzdr0KBBqlmzpsqUKaMKFSqoa9euWrp0qaZOnXpZ/Xbr1k1RUVF67LHHVKdOHTmdTpUrV07t2rXTJ598ogULFhTo6hAAAAAAAMA+j61McGnevLnmzZtnq01el310CQsL0wcffHC5w8phyJAhGjJkSKH1BwAAAADAtYqP7wEAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsIUwAQAAAAAA2EKYAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGzxKe4BAABwqdpjI4t7CAAAAMgDKxMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsIUwAQAAAAAA2EKYAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBaf4h4AAADwvNpjI6+4j+hJ4YUwEgAAcC1gZQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsIUwAQAAAAAA2EKYAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFt8insAAIBrS+2xkcU9BAAAAHgYKxMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANji8TBh27ZtGjRokGrWrCmn06mQkBD16tVLX3/99RX1Gx0drccff1x169ZV2bJlVaVKFXXu3Fnz58/Pt21ycrKmTJmi22+/XRUqVJDT6VS1atXUvXt3zZ8/XxkZGVc0NgAAAAAArmUevTTk8uXL1bdvX6WlpVm3HT16VCtXrtTKlSs1atQoTZs2zXa/GzZsUJcuXZScnGzdlpCQoDVr1mjNmjVatGiRFixYIB+fnIe3c+dOhYeH68CBA9luP378uL7++mt9/fXXmjt3rpYsWSJ/f3/bYwMAAAAA4FrnsZUJW7ZsUf/+/ZWWlqbWrVtr3bp1SkhI0MaNG3XPPfdIkt59913NmDHDVr9xcXEKDw9XcnKyGjRooJUrVyo+Pl7bt2/XsGHDJElLlizRiy++mKNtSkqKunfvrgMHDsjPz08TJ07UX3/9pWPHjunHH3+0xrV69WoNHTr0io4fAAAAAIBrlcfChFdeeUXnzp1T/fr1tWbNGnXo0EGVKlVSq1attGTJEvXr10+SFBERkW2FQX4mTZqkhIQEBQcHa/369QoPD1flypXVtGlTffzxxxozZowkadq0aYqOjs7WdsaMGTp48KAkadmyZRo7dqwaNGigqlWrql27dvryyy/19NNPS5IWLFig33///cofCAAAAAAArjEeCRN27dqlyMhISdK4ceMUGBiYrdzhcGjq1Kny8vLSiRMntGTJkgL1m5iYqNmzZ0uSRo0apZCQkBx1IiIiFBwcrLS0NM2dOzdb2aJFiyRJHTt2VNeuXd3ex/jx462vR7iOAQAAAAAA/I9HwoRVq1ZJygwNevXq5bZOaGioWrRoIUlaunRpgfpdu3atzp8/L0nq3bu32zqBgYHq3Lmz235PnDghLy8vtW3bNtf7CAoKUpUqVSRJhw8fLtC4AAAAAAAoTTwSJmzdulWSFBYWpsqVK+dazxUmbNq0yVa/Pj4+uvHGG/Ptd9u2bUpNTbVu37dvny5cuKBXX30117ZJSUmKj4+XJFWoUKFA4wIAAAAAoDTxyNUcXHsV1KlTJ896YWFhkjI3Vbx48aLbqy+46zc0NFTe3t759puenq7Y2FjVq1fPKvPx8cnzfmbNmqWLFy9Kktq1a5fneNyJi4vLszwlJcV2nwAAAAAAlCQeCRMSEhIk5f/JflBQkCTJGKPExMQ8VzFcTr+SdOrUqXzH67J3715NmDBBklS/fn1169atwG1dQkND8ywvX768Ro8ebbtfAAAAAABKCo98zcG1r4Gfn1+e9bKWu9oUR7+SdOzYMfXs2VNJSUlyOByaPn26ypQpU6C2AAAAAACUJh5ZmZDXVxBKYr+HDx9W165dtXv3bkmZV3S4nFUJkhQbG5tneUpKir744ovL6hsAAAAAgJLAI2FCQECApPxXBZw7d876Pb/VBp7qd+fOnerevbsOHjwoSXrmmWfy3KAxPzVr1syzPCkp6bL7BgAAAACgJPDI1xyCg4MlSadPn86zXmJioqTMFQcFuXKC3X4l5bkPw3fffadbb73VChJeeeUVvfXWW/mOAwAAAACA0swjYULDhg0lyXqTnpuYmBhJUo0aNeTllf9QXP3GxsbKGJNvvz4+PgoJCXFb59NPP1WPHj10+vRpeXt768MPP9Rrr72W7xgAAAAAACjtPBImNG/eXJK0f//+PJf1b968WZLUokULW/1euHBBUVFR+fbbtGlTOZ3OHOX/+Mc/NHToUKWlpSkwMFDLli3TiBEjCjQGAAAAAABKO4+ECT169JAkpaenKzIy0m2d2NhYbd26VZJ01113Fajfjh07yt/fX5K0fPlyt3VSUlK0Zs2aXPt9//339cILL0iSqlatqnXr1ik8PLxA9w8AAAAAADwUJtStW1ft2rWTJEVEROTY48AYo2effVYZGRmqXLmyBg0aVKB+AwMD1adPH0nS1KlTra8zZDV+/HglJibK6XTqiSeeyFa2YcMGPf3005KkKlWq6Mcff1TLli3tHh4AAAAAAKWaR8IESXr77bfl5eWlPXv2qH379vrmm2+UkJCgzZs367777tPChQslZb75d12lwaVx48Zq3LixHnrooRz9vvnmmwoICNCJEyfUvn17LV68WPHx8dq5c6dGjBihKVOmSJJGjRqV48oKI0eOVFpamhwOh2bOnKnq1avrzJkzuf5cuHDBQ48OAAAAAABXL49cGlKSWrVqpdmzZ2vYsGHatm2bunXrlqPO6NGjNXLkyBy37969W5J03XXX5SgLDQ3V4sWL1adPH8XExKhv37456vTr10+TJ0/OdttPP/2k33//XVLmyoh77rkn32MYPHiw5syZk289AAAAAABKE4+tTJCkIUOGaPPmzRo0aJBq1qypMmXKqEKFCuratauWLl2qqVOnXla/3bp1U1RUlB577DHVqVNHTqdT5cqVU7t27fTJJ59owYIFOa4O8d///rcwDgkAAAAAgFLPYysTXJo3b6558+bZapPXZR9dwsLC9MEHHxS4zzFjxmjMmDG2xgEAAAAAAHLy6MoEAAAAAABw7SFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgi8cvDQkAAK4NtcdGXnEf0ZPCC2EkAACguLEyAQAAAAAA2EKYAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGCLT3EPAABQctQeG1ncQwAAAMBVgJUJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsIUwAQAAAAAA2EKYAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbfIp7AACAwlN7bGRxDwEAAAClACsTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbfIp7AAAAoPSoPTbyivuInhReCCMBAABXgpUJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANjiU9wDAABkqj02sriHAAAAABQIKxMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsMWnuAcAANeC2mMji3sIAAAAQJFhZQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGzxeJiwbds2DRo0SDVr1pTT6VRISIh69eqlr7/++or6jY6O1uOPP666deuqbNmyqlKlijp37qz58+fn29YYo3nz5qlDhw4KCgqSn5+fGjRooKefflqxsbFXNC4AAAAAAK51DmOM8VTny5cvV9++fZWWlua2fNSoUZo2bZrtfjds2KAuXbooOTnZbXmfPn20YMEC+fjk3F8yIyNDAwYM0BdffOG2bVBQkL788kt16tTJ9rgKIikpSW+//bYk6ZlnnlH58uU9cj8AihYbMAJXj+hJ4cU9BAAAipQn3od6bGXCli1b1L9/f6Wlpal169Zat26dEhIStHHjRt1zzz2SpHfffVczZsyw1W9cXJzCw8OVnJysBg0aaOXKlYqPj9f27ds1bNgwSdKSJUv04osvum3/0ksvWUHC6NGjtXv3bh07dkz/+c9/FBoaqtOnT6tPnz6Ki4u7/IMHAAAAAOAa5rGVCT179lRkZKTq16+vLVu2KDAw0Cozxuhvf/ubFi5cqEqVKunAgQMqV65cgfp94oknNGPGDAUHBysqKkohISHZyp977jlNmTJFZcqU0V9//aXatWtbZYcOHVLdunWVmpqqF154QZMmTcrW9uDBg2rZsqVOnDihYcOG6eOPP778ByAXrEwASh5WFQClCysTAACljSfeh+b8HkAh2LVrlyIjM1+cjxs3LluQIEkOh0NTp07V4sWLdeLECS1ZskSDBw/Ot9/ExETNnj1bUuZXJC4NEiQpIiJCs2bNUmJioubOnauIiAirbPr06UpNTVVgYKBeeumlHG3DwsL0zDPP6OWXX9bnn3+ud955R/7+/raOHQAAoKgQhmZHUAQARccjYcKqVaskZYYGvXr1clsnNDRULVq00KZNm7R06dIChQlr167V+fPnJUm9e/d2WycwMFCdO3fW4sWLtXTp0mxhwldffSVJ6tSpU64rIXr37q2XX35ZZ8+e1bfffpvr/QAAAFwJgoDCVxiPKYEEABSMR8KErVu3Ssr8pL9y5cq51nOFCZs2bbLVr4+Pj2688cY8+128eLG2bdum1NRUOZ1OpaWlKSoqSpLUsmXLXNs2bdpUTqdTqamp2rRpE2ECAADXGN7EIy8EEgBQMB4JE6KjoyVJderUybNeWFiYpMxNFS9evOj26gvu+g0NDZW3t3e+/aanpys2Nlb16tWz7iO/cTkcDoWGhmrfvn06cOBAnuNxJ7+NG7NegSK3q1EAKLhb3vz+ivvgy0wAgMJ0/dgvr7iP/47rXAgjAYBMWd97ZmRkFEqfHgkTEhISJEkVKlTIs15QUJCkzA0ZExMT81zFcDn9StKpU6eytbXT3tXWjtDQ0DzLq1evruHDh0uSZs2aZbt/ANn9za+4RwAAQOF7++0/i3sIAK5RKSkpCg4OvuJ+PHJpSNe+Bn5+eb/Kz1ruauOpfrP2X9D2BRkTAAAAAACljUdWJuT1FYTi6tdTY7pUbGxsnuUXL17UqVOnVLVqVQUEBMjLyyN5TpE6cuSI2rRpI0nasGGD26tsoORjHq9+zOHVjzm8+jGHVz/m8NrAPF79mMPClZGRoZSUFElStWrVCqVPj4QJAQEBkvL/ZP/cuXPW7/mtFrjSfl1t7bQvyJguVbNmzXzr1K5d23a/JVlSUpKSkpIkZT7OhXHNUhQ95vHqxxxe/ZjDqx9zePVjDq8NzOPVjzksfIXx1YasPPKxuGuQp0+fzrNeYmKipMxVA/ntY3A5/Uqy9mHI+sAVtH1+ezgAAAAAAFAaeSRMaNiwoSTp4MGDedaLiYmRJNWoUaNAy/1d/cbGxsoYk2+/Pj4+1nKY0NBQa6VBXuMyxlhXZKhVq1a+YwIAAAAAoLTxSJjQvHlzSdL+/futpSnubN68WZLUokULW/1euHBBUVFR+fbbtGlTOZ1OSZKXl5euv/56SdKWLVtybbt9+3alpqbaGhcAAAAAAKWJR8KEHj16SJLS09MVGRnptk5sbKy2bt0qSbrrrrsK1G/Hjh3l7595Rfjly5e7rZOSkqI1a9a47dc1ru+//97afOJSrn7Lli2rjh07FmhcAAAAAACUJh4JE+rWrat27dpJkiIiInLsUWCM0bPPPquMjAxVrlxZgwYNKlC/gYGB6tOnjyRp6tSp1tcZsho/frwSExPldDr1xBNPZCsbOHCgvL29derUKU2YMCFH25iYGL399tuSpKFDhxb6BhUAAAAAAFwLPHZdwrffflteXl7as2eP2rdvr2+++UYJCQnavHmz7rvvPi1cuFBS5pv/rFdakKTGjRurcePGeuihh3L0++abbyogIEAnTpxQ+/bttXjxYsXHx2vnzp0aMWKEpkyZIkkaNWpUjisrNGzYUCNHjpQk/fOf/9Rjjz2mnTt3Kj4+XosXL1b79u114sQJVaxYUWPHjvXEwwIAAAAAwFXPI5eGlKRWrVpp9uzZGjZsmLZt26Zu3brlqDN69GjrzX1Wu3fvliRdd911OcpCQ0O1ePFi9enTRzExMerbt2+OOv369dPkyZPdjmvy5Mnat2+fIiMj9dFHH+mjjz7KVh4QEKCVK1ey+SIAAAAAALlwmLwui1AItm3bpn/+859au3atjh07psDAQLVq1UojR45U79693Q/K4ZAkdejQQevWrXNb5+DBg5o0aZJWr16tQ4cOqWzZsrrxxhs1dOhQDRkyxOrDHWOM5s2bp08//VR//PGHUlJSVL16dXXr1k0vvPCC6tate8XHDQAAAADAtcrjYQIAAAAAALi2eGzPBAAAAAAAcG0iTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsIUwAQAAAAAA2EKYgMv2119/acSIEapXr558fX1Vvnx5tWzZUq+//rqSkpLybb9ixQrdddddqlSpksqWLavatWvr0UcfVVRUVBGMHi4ZGRmaM2eOunTpoqpVq8rpdKpGjRp68MEHtWnTpjzbGmM0b948dejQQUFBQfLz81ODBg309NNPKzY2toiOAO4MGDBADodDHTt2zLcu52Lx27Fjhx577DE1atRI/v7+8vf3V4MGDTR8+HDt2LEj3/bMYcmwbds2DRo0SDVr1pTT6VRISIh69eqlr7/+uriHhixWrVqlvn37qmbNmipbtqzKly+vm2++Wa+88ori4+NzbZeWlqZ3331Xbdq0UWBgoAICAtS0aVO9/PLLOnnyZBEeAS6VkpKihg0byuFwaPz48bnWYw5LhuTkZL3++utq2bKlgoOD5evrq0aNGumZZ57R4cOH82ybkpKiv//977rhhhvk7+9vnb+TJ0/WuXPniugIYDHAZVi4cKHx8/Mzktz+1K1b1+zZsyfX9s8//3yubcuWLWs+//zzIjya0uvkyZPm9ttvz3UuvL29zYcffui2bXp6uunfv3+ubYOCgsyaNWuK+IhgjDHz58+35qFDhw551uVcLH7vvvuu8fHxyXUefHx8zAcffJBre+awZFi2bJkpU6ZMrnMxatSo4h5iqZeWlmYGDBiQ6xxJMlWrVjW//PJLjrbnzp0zHTp0yLVd9erVzbZt24rhqGCMMSNGjLDmIiIiwm0d5rBk2Lp1q6levXqu81CxYkXz22+/uW2bkJBgmjRpkmvbxo0bm0OHDhXxEZVuhAmwLSoqyjidTiPJ1KhRw3z22WcmJibG7Ny500yYMMGULVvWOqEvXLiQo/2HH35onfSDBg0yf/75p4mPjzerVq0yzZo1M5KM0+k0mzdvLoajKz0yMjJMu3btjCTjcDjMM888Y6KiokxcXJxZuHChCQsLM5KMl5eX2bBhQ472Y8eOteZx9OjRZvfu3ebYsWPmP//5jwkNDTWSTHBwsImNjS2Goyu9YmNjTYUKFQoUJnAuFr/ly5dbc9C8eXPz5ZdfmqNHj5r9+/ebf//739Z56HA4TGRkZI72zGHJsHnzZitgb926tVm3bp1JSEgwGzduNPfcc481R9OnTy/uoZZqzz77rDUXvXv3Nj///LNJSEgw27ZtM5MnTzYBAQHWm5m4uLhsbV3heZkyZcwbb7xhDhw4YA4fPmxmzpxpPefWrl3bnDlzppiOrvRauXJltjeUuYUJzGHxO3LkiKlUqZL1odP06dNNdHS02bNnj3nrrbeMv7+/FewkJSVla5uenm5uu+02I8mUK1fOvP/+++bQoUPm4MGD5h//+Ifx9fU1kkybNm1Menp6MR1h6UOYANseeeQRI8n4+/ubXbt25SifOXOm9YR+6SdiKSkppkqVKkaS+dvf/paj7alTp0z9+vWNJNO1a1ePHQOM+eijj6x5crf64NChQ6Z8+fJGkunbt2+2sri4OCtQeuGFF3K0jY6Otv5YDBs2zGPHgOwyMjJM586ds72oyi1M4FwsGa6//nojyTRq1MgkJyfnKD958qSpU6eOkWSaNm2arYw5LDnCw8ONJFO/fv0c85iRkWH69etnJJlKlSrleIGMonHo0CFrBdCAAQPc1tm4caNVZ+TIkdludz2nulsltHnzZutv4htvvOGxY0BOx48fN9WqVcs3TGAOS4YHHnjASDKBgYFm48aNOcqzBkOXhq8LFy60ylatWpVnW1bkFR3CBNjWvHlzI8l0797dbXlaWpqVDl66rDPrp2j79u1z2/6zzz6z6hw4cKCwh4//r2nTpkaS6datW651hg8fbry8vEzDhg2z3e5alRAYGJjrC+PXX3/dCp1SUlIKdexw76233jKSTJMmTUzbtm3zDBM4F4vfjh07rMd3zpw5udabMWOGVS86Otq6nTksGXbu3Gk9xp988onbOjExMcbLyyvfuYbn5HYeXcoV/NSuXdu6zfWJdp06dczFixfdtnv00UetOig6vXv3NpLMkCFD8gwTmMPid/ToUePt7W0kmYkTJ+Zar2HDhqZMmTJm6NCh2W6/5ZZbjCRzxx135Nq2S5cuRpLp2LFjoY0beWMDRtjm5ZX5zyYtLc1tucPhkLe3tyRZ/3X56quvJEnNmzdX3bp13bbv2bOn1W7ZsmWFMmZkt2PHDmtTtxdeeCHXeu+9955SU1O1e/fubLe75rFTp04qV66c27a9e/eWJJ09e1bffvttYQwbedixY4fGjRsnHx8f/etf/5Kvr2+e9TkXi9+BAwdUvnx5SVLbtm1zrVe/fn3r96wbUzGHJcOqVaskZf7t69Wrl9s6oaGhatGihSRp6dKlRTU0ZHH48GH5+fmpWrVqCgsLy7We63xznWvGGGsDzazn06Vcf/MOHDigP/74ozCHjlzMnj1by5YtU1hYmKZNm5ZrPeawZFi0aJHS09Pl7++vJ598Mtd6f/zxh1JTUzV79mzrtpMnT+q3336T9L95csdV9sMPP+jUqVOFNHLkhTABtrVp00aS9PPPP2vPnj05yv/zn/8oJSVFknT77bdnK9u6daskqWXLlrn2HxQUpDp16khSvlcTwOXZsGGDJMnpdOaYo6whkdPpzPFHNy0tzdohPq95bNq0qZxOpyTm0dNSU1M1cOBAnT9/Xi+//HKe8+LCuVj8wsPDdfr0aSUnJ6tBgwa51tu7d6/1e4UKFazfmcOSwTUPYWFhqly5cq71XGEC81A8Xn/9dZ09e1Z//fVXnvVc55vrXIuOjlZiYqKkvM811/xKzHFR2Ldvn55++mk5HA7NmTPHCmbdYQ5LBtdrzzZt2iggICBbWdbXnu4+DPnjjz9kjJFUsDnMyMjQli1brnjMyB9hAmwbO3asKlasqHPnzqlr16764osvFBcXp71792rKlCkaNmyYJKlLly669957rXbp6emKi4uTJOvFbW5cnxocOHDAQ0dRum3fvl1S5uPsdDr1559/auDAgapSpYqcTqeCgoJ07733auPGjTnaxsXF6eLFi5LynkeHw6HQ0FBJzKOnvfLKK9q6datat26tl156Kd/6nIslS2BgYK6flGVkZOjjjz+WJFWtWlUNGzaUxByWJNHR0ZIKPg9Zn0NR9PJ603n48GGtWLFCktSuXTtJ/5tfKe85DgkJUZkyZSRxrnlaenq6Bg0apDNnzuipp57K9xLIzGHJ4Hrt6QrPly9frq5duyowMFBOp1PVq1fX//3f/7m9NGRB5zDrqiPmsGgQJsC2unXr6ocfflCnTp108OBBPfDAAwoNDVWDBg303HPPyRijV199VStWrLC+EiFlLlHKyMiQlP3TNXeCgoIkiSVKHnLkyBFJUqVKlTRnzhy1bt1an3/+uRISEiRJSUlJWrp0qW699VZ98MEH2dq66kjMY0nwww8/aMqUKfLz89O8efPk4+OTbxvOxavHlClTrOW2jz32mPWcyhyWHK7nxILOgzHG+pQUJYcxRsOHD9f58+clSSNHjpRU8L95Xl5e1tf+ONc8a+LEifr111/VpEkTTZw4Md/6zGHJ4HrtWbFiRY0YMUK9e/fWd999Z61mPnLkiD744APdcMMN+vXXX7O1Legcup5nJeawqBAmlEKvv/66HA6HrZ8hQ4Zk6+P06dMKDAyUw+HI0f/Zs2e1cePGHN+zd/2BliQ/P788x+gqz9oG2V3JPCYnJ0vKXCY4bNgwXXfddVqwYIESExN15swZLVu2TPXr11d6erpGjhypb775xrpf5rHwXOm5mJSUpMGDBysjI0MTJ05U48aNC3S/zGHhKYzn09wsXrxY48aNkyQ1bNhQzz33nFXGHJYcrse1oPOQtQ1KjtGjRysyMlKS9OCDD6pTp06SONdKmk2bNum1114r8P5AEnNYUrhee86bN08ff/yx2rdvrx9//FHnzp3T8ePHNX36dAUGBurEiRPq3bu3FT5IBZ9DnmeLHmECbFuwYIE6duyoFStWqFOnTtYTwalTp7RkyRI1atRIq1atUvv27fXzzz9b7XJbxouid/bsWUlSfHy8KlWqpF9++UX333+/goKCFBAQoLvvvls///yzqlevLmNMtjcxzGPJ8eSTTyo6Olp33nmnRo0aVeB2zGHJt3DhQj3wwANKT09X+fLltXjxYgUGBlrlzGHJwVxc3YwxGj16tN555x1JmRuafvTRR1Y581tynDt3TgMHDlRaWlqB9weSmMOSwvXa88iRI+rUqZO+//57tWvXTr6+vqpSpYpGjhypyMhIeXl5KT4+XpMmTbLaMoclV/7rYXHNefzxx9W3b19bbVzLhk6ePKnhw4crLS1N4eHhWr58ubXs1tfXV/fee6/uvPNO3XbbbYqKitLQoUMVFRUlb2/vbJut5JcWnjt3TlL+CXJpdiXz6O/vb932wgsvqEaNGjnqVq1aVc8++6yeffZZ/fnnn9q7d6/q16/PPBaiK5nDRYsWad68eQoKCtKnn37qdpVQbpjDwnMlc5ib6dOn66mnnlJGRobKlSunyMhINWvWLFsd5rDkcM1FQedBYi5KitTUVD3yyCP67LPPJElNmjTRN998ky2441wrOZ577jnt2rWrwPsDuTCHJYO/v7/OnDkjSZo6daq1P0VWd9xxh8LDw7VixQotXrzYukrHpXOY9RzNiufZokeYUApVqlRJlSpVuqy2CxYsUFJSkiTp7bffzrYngktQUJAmT56sXr166a+//tKaNWvUtWtXlStXTt7e3kpPT9fp06fzvB/X90nz2hm7tLuSecx6OUfXUk53OnToYP0eFRWl+vXrKzg42LqNebwylzuHR44c0YgRIyRJ06ZNU61atWy151wsPFdyHl4qIyNDzzzzjN59912r76+++sq6gk5WzGHJ4XpOLOg8eHt757u/Ajzv5MmTuvfee/XDDz9IytwhftWqVapSpUq2egX9m5eRkWEt4+ZcK3yrV6/WjBkz5OvrW+D9gVyYw5KhXLlyOnPmjIKCgrJdOeNSHTp00IoVK3To0CElJSWpfPnyOeYwtzAh6340zGHR4GsOsMV1SaUqVarkeSmzrDvr7tq1S1Lmxjb16tWTJB08eDDP+4mJiZEk22+SUDBZd8LN6/uGWXe9dqW9oaGhVtqb1zwaY6zd5pnHwrV69WqdPHlSkjRkyBC338tfv369JGn9+vXWbXPmzJHEuVgSnT17Vvfee68VJNStW1e//PKL2yBBYg5LEtcVNgo6DzVq1HAbxKPo7Nu3T7feeqsVJNx1111at25djiBB+t/8SnnP8ZEjR6zL23GuFb758+dLyvxUukmTJm7/7rlMmDDBui06Opo5LCFcrz3z2+fC3WvPgs6h63lWYg6LCn/NYEtqaqok6cKFCwVuk7Vu8+bNJSnPa7+ePn3aupxLXsklLt9NN91k/b5///5c6x07dsz63fVVCC8vL11//fWS8p7H7du3W/9emMeSh3Ox5EhOTtadd96p5cuXS5Latm2rX3/9NduLJ3eYw5LBNQ/79++3Vu65s3nzZknMQ3HbsWOHbr31VuvDkWHDhmnFihW5ftIZEhJirT7K61xzza/EHJc0zGHJ4HrtGR8fb60Accf12rNMmTJWwNe0aVMrMCrIHDocDt14442FMWzkgzABtjRq1EhS5i7yGzduzLXejz/+aP3epEkT6/cePXpIynwicH1qfamVK1cqPT1dktStW7crHjNy6tKli5xOp6TM797nxnUVBz8/v2xPyq55/P77761L+lzK9caobNmy+V4DGvYMHDhQycnJef64rpHerl0767aBAwdafXAulgwXLlxQjx499Ntvv0mSevXqpbVr16pq1ar5tmUOSwbXPKSnp1tXA7hUbGystm7dKinzU3AUj/3796tLly6Kj4+XJP3973/Xxx9/nO+Sedccr1y5UsYYt3Vcf/NCQkJ4E+MBH330Ub5/91xefPFF67awsDBJzGFJEB4eLinz6yRffvllrvVcrz3btm1rreIqX7689brGNU/uuMratGmjihUrFsq4kQ8D2BAbG2ucTqeRZG6//XZz/vz5HHWSk5PNDTfcYCSZatWqmbNnz1plp06dMoGBgUaSuf/++3O0PXXqlKlfv76RZLp37+7RYyntBgwYYCQZb29vs379+hzle/fuNUFBQUaSGThwYLay3bt3G29vbyPJPPfccznaHjx40FSqVMlIMo8//rjHjgG569Chg5FkOnTo4Lacc7FkGDVqlJFkJJn77rvPpKWlFbgtc1hytGvXzkgyDRo0MImJidnKMjIyTL9+/YwkU7lyZXPmzJliGmXplpqaalq1amWdb2+//XaB23733XdWuxkzZuQo37x5s/XaaPLkyYU4atjhmqOIiIgcZcxh8UtLSzNhYWFGkqlZs6Y5evRojjoLFy605mnWrFnZymbNmmWVrVy5MkfblStXWuULFizw2HEgO8IE2DZhwgTrZL3pppvM4sWLzeHDh83Ro0fNwoULzfXXX2+Vz58/P0f7t956yyrv06eP2bx5s4mPjzerV682zZo1M5KMr6+v+f3334vh6EqPI0eOmMqVKxtJxs/Pz7z55ptmz5495tixY+bzzz831atXN5JMcHCwiYuLy9E+65ugESNGmKioKHP8+HGzaNEiU6tWLSPJVKxY0Rw8eLAYjg75hQnGcC4Wtz///NM4HA4jydSvX98cPXrUJCcn5/lz8eLFbH0whyXDxo0bjZeXl5FkmjdvblavXm3i4+PNpk2bzL333mvN0fTp04t7qKXWe++9Z83D/fffn++5lpycnK393XffbQXw48aNM/v27TNHjhwxs2fPNhUrVjSSTJ06dUxSUlIxHSHyChOMYQ5LgtWrV1vPlWFhYeZf//qXOXTokImOjjZvvPGGFejccsstOf7eXbx40bRo0cJ63TplyhQTGxtrYmNjzZQpU4yfn5+RZNq2bWvS09OL6QhLH8IE2JaRkWHGjBljPWm7+ylTpox5//333bZPT083w4YNy7Wtj4+PWbJkSREfVen0559/mtDQ0FznonLlym5XLRhjzLlz50x4eHiubQMCAswvv/xSxEcEl4KECZyLxeuRRx7J83nU3c/atWuz9cEclhyffvqp8fHxyXUuRo8eXdxDLNXq1atn+3zL6uTJk6Z169a51q1WrZrZs2dPMR0djMk/TGAOS4b58+dbb/zd/dx8880mNjbWbdvo6GhTt27dXNs2atTIHD9+vIiPqHQjTMBl27Bhgxk8eLCpU6eO8fX1Nf7+/qZx48Zm5MiRZvfu3fm2X7ZsmbnrrrtM5cqVjY+PjwkJCTEPPvig2bp1axGMHi5nzpwxEydONK1atTLly5c3AQEB5vrrrzcvvviiOXToUJ5tMzIyzJw5c0yHDh1McHCwKVOmjAkLCzPDhw83+/btK6IjgDsFCRNcOBeLR9OmTa84THBhDkuGP//80wwaNMjUrFnTlClTxlSoUMF07drVLF26tLiHVqrFx8fbPtcuDROMyfyqxLRp00ybNm1MuXLljNPpNA0aNDCjR482x44dK4YjQ1b5hQnGMIclxcGDB81TTz1lGjZsaPz8/EyFChXMLbfcYqZPn57t69HuJCcnm9dee83ccMMNJiAgwPj6+ppmzZqZ8ePH51hRBM9zGJPLLiQAAAAAAABucDUHAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC2ECQAAAAAAwBbCBAAA4HHp6enFPQTkgfkBANhFmAAAgAesW7dODodDDodD69atK+7hXJErPZYvvvhCAwcOzHH7+PHjrX5Ls+joaOtxmDNnju32Q4YMkcPhUO3atd2Wf/7552rbtq2CgoLk7++vevXq6bvvvpMkHT16VP3799dPP/10BUcAACiNCBMAAIDHvPTSS3rggQd05MiR4h5KqfTJJ59o4MCB2rBhg5KSknTu3Dnt379f1atXV3x8vJo0aaIFCxbIGFPcQwUAXGV8insAAADg2nXo0KFcyypWrKh69eoV4WiuTdWqVVO9evVUs2bNHGWffPKJJKlWrVpasGCBGjZsqDNnzqhmzZqKiYlRYmJiEY8WAHCtYGUCAAAoFqNGjdLevXu1d+/e4h7KVW3y5Mnau3ev26+gHD16VJLUr18/3XLLLapYsaJq1aolLy9eAgIArgx/SQAAAK5RFy9elCQFBgYW80gAANcawgQAAIrJn3/+qWHDhql+/fry8/NTUFCQbr75Zr3yyis6ceJEnm0vXLiguXPnqnPnzqpevbqcTqdq1Kih/v3767///W+u7Xbs2KGnnnpKN910kypWrKgyZcqoUqVKatu2rcaPH6+TJ08WyrG5NlecO3euJGn9+vXWJoPR0dHZ6ly6AWPWDR8laevWrXrggQcUEhIiPz8/NWjQQOPGjVNKSoqkzCsRvPfee7r55psVEBCgoKAgde7cWd9//32eY/z11181cOBA1apVS76+vqpYsaI6dOigDz/8UGlpaZd97KtXr1a/fv1Uo0YNOZ1OVahQQS1bttS4ceOslQJ5OXLkiJ566inVq1dPvr6+uu666xQeHq7Vq1e7re9uA0bX43fw4EFJ0oQJE7Jt8uhwOFSnTh2rfqdOneRwODRkyJDLPm4AQCljAABAoVu7dq2RZCSZtWvX5iifMGGC8fLysupc+lOhQgXz3Xffue07JibGtGrVKte2DofD/OMf/8jRbvz48cbhcOTaTpIJDQ01MTExto7FnYiIiFzv48CBAznq5HZ/CxYsME6n020/t956qzlz5oy588473ZZ7eXmZpUuX5hhbenq6GT16dJ6PQ4sWLcyhQ4cKdKxZvfTSS3n2W758efPLL79ka3PgwAGrfNiwYSY4ODjX9pMnT85xn4MHDzaSTFhYmHVbXmP49NNPcy0bPHiw7WMGAJROrEwAAKCIvfXWW4qIiFBGRoZuvvlmrVixQseOHVNMTIw++OADValSRadOnVLPnj21devWbG3T0tLUo0cP/f777/Lx8dG4ceMUFRWl48eP69tvv1XLli1ljNHzzz9vXf5PkhYtWqTx48fLGKOuXbtqzZo1OnTokA4dOqQ1a9YoPDxckhQbG6uIiIgrPsZx48YpOTlZAwYMkCS1a9dOycnJSk5OVlhYWIH7GTx4sGrUqKGFCxfqyJEj2rJlizp27Cgpc2VBy5YttW7dOj3//PPatWuXjh8/rvnz56tChQrKyMjQ888/n6PPiIgIvfXWW5KkPn366KefftKJEye0b98+TZ06VeXLl9eWLVsUHh6u8+fPF3isv/76q9544w1J0oABA7Rx40YdP35c+/bt0/vvv6/AwEAlJSXp4YcfVkZGhts+Zs6cqbS0NGsfhNjYWM2dO1eVKlWSlHl1jILsMeF6rGvVqiVJevHFF63bBg4cqOTkZO3YscOq/9VXXyk5OVkfffRRgY8XAFDKFXeaAQDAtSi3T/OPHj1q/P39jSTTrl07c+7cuRxt9+zZYypUqGAkmTZt2mQre+utt7J9an+pU6dOmbCwMCPJ3HnnndbtrVu3NpJM06ZNzfnz53O0S09PNzfffLORZEJCQgp0LAXh+tS8Q4cOOcoKsjKhXLlyJi4uLlv5oUOHjI+Pj1Vn0qRJOfp+//33rfKDBw9at//111/WipAnn3zS7Zg3bNhgvL29jSQzderUAh/rmDFjjCRTv359k5GRkaN8zpw51pg2bdpk3Z51ZYKXl5f58ccfc7Rdvny5Veedd97JVuZuZYKL699CREREjrKs92t3XgEAYGUCAABF6N///rfOnj0rSfrggw/k6+ubo079+vWt1QEbNmzQpk2bsrWXpI4dO+r+++/P0TY4OFhPP/20WrZsqdDQUBljlJGRoZ49e+qhhx7Sq6++qrJly+Zo5+XlpTvuuEOSFB8ff+UHWkgGDhyoGjVqZLutevXq1iUlfX19NWrUqBztbr31Vuv3rJen/Pjjj5WRkSF/f3+9+eabbu+zdevW6t+/v6TMlQIFdeHCBUnSmTNndObMmRzl/fr107Jly7Rt2zY1bdrUbR/du3dXu3btctzeo0cP69/K/v37CzwmAAA8hTABAIAitH79eklSo0aN1KxZs1zrZQ0KXG1Onz5tBQs9e/bMte3TTz+t33//3dpoz8vLS6+++qrmzp3rNoDIyMjQ9u3bdeDAAUn/uwJASdC2bVu3t1etWlWS1LhxY/n5+eUoL1++vPW7602+JOvyiU2aNJH0vzf+l/647nfXrl35bobp4gpjjh49qpYtW2rq1KnatWuXVe7v76+7775bzZo1cxvoSNJtt93m9nZvb29VrlzZGjMAAMXNp7gHAABAaRIbGyvpf29mcxMSEqLg4GAlJiYqJiZGknT48GEZYyRJDRs2vKz7P3bsmL7//ntFRUVp37592rt3r3bu3GldGaGkcb2BvpSXV+bnIUFBQXmWX8r1qf6mTZtUrly5Ao0hNjbW2rMgL3369NE999yjpUuXas+ePRozZozGjBmjWrVqqVu3burVq5e6desmp9OZax+5Ha+UGShImVevAACguBEmAABQhJKSkiRJgYGB+dYNCAhQYmKi9Ul01ss2+vv727rf8+fP65lnntHMmTNzvBn19fVVp06dlJ6erh9++MFWv54WEBBQqP25Hn9PtPHy8tLixYv1ySef6MMPP7RWkcTExGjmzJmaOXOmqlSpoqlTp2rQoEFu+8graAAAoCThaw4AABQhV4hQkKXqycnJkv73hjrrG2vXvgsF1b9/f3344YdKT09Xy5Yt9corr2jBggXatm2bkpOTtWbNGrVv395Wn1cjVwjzt7/9TcaYAv24vr5QEF5eXnr00Uf1+++/KzY2VrNmzVL//v2tlQ3x8fF66KGHtGLFCo8cHwAARYUwAQCAIuS6LOLOnTvzrHfo0CHrE3FXm5o1a1rl+/bty7VtTEyMXnzxRc2cOVMnTpzQL7/8omXLlkmSRo0apd9//12vvfaa7r//fjVr1kw+PpkLFRMSEi7/wK4SrkslRkdH51nP9XWSK1GzZk098sgjmj9/vo4dO6Z///vf1l4J77777hX3DwBAcSJMAACgCLk+/d+9e7e2b9+ea72FCxdav7s25atcubIaNWokSfr6669zbbtixQpNmjRJw4cPV1pamn755RerbPjw4W7bZGRkaO3atdn+vzA4HI5C6aewuFYZbNq0SXFxcbnWGzFihCpVqqTWrVtbK0Tyc//996tu3bp64YUXcpR5e3vrgQceULdu3SRlv8JEcSlpcwMAuLoQJgAAUIQeeugh6xJ/jz/+uM6fP5+jzv79+/X3v/9dktSsWbNsVzR4+OGHJUmrV692GygkJSVpypQpkqQOHTrouuuus1YeSFJUVJTbcU2YMEF//fWX9f+pqal2D80t130XVn9XatiwYZIyr1gxcuRIt5sZ/vbbb5o7d65OnjypihUrFnijxrNnz+rAgQP67LPP3K7ySE1NtR5/16Uti1PWfxclZX4AAFcPwgQAAIpQlSpV9MYbb0iSfvrpJ7Vr106RkZGKj49XXFycZs6cqdtuu00nT56U0+nUvHnzsn2C/OSTT1pXgujTp48mTpyo/fv36+jRo/rqq690xx13KDo6Wt7e3po8ebIkqWvXrlYfTzzxhD7//HPFxcXp8OHDWr16tXr16qXXXnst2zgL+ml8flx7BWzbtk2bN2/WiRMnlJaWVih9X46bbrpJ//d//ydJWr58uTp16qRvvvlGCQkJ2rdvn9577z11795dqamp8vX11T//+c8C9/3cc89JyrzqRqdOnbRkyRIdPHhQx44d0/r169WzZ0/t3btXkqwxFKeKFStavy9cuFAnT57UqVOninFEAICrCWECAABFbPTo0Ro/frwcDoc2bdqknj17qmrVqgoNDdXw4cN17NgxValSRZGRkWrRokW2tv7+/vrqq6/UpEkTnTt3TuPGjVO9evUUEhKi8PBw/fHHH/L19dW8efOsFQ1Nmza1lt4fP35cAwcOVGhoqGrUqKG77rpLK1euVFBQkJ599lnrfvbs2VMox9qpUydJmRtOtmzZUpUrV9bPP/9cKH1frnfeeUePPPKIJOnHH39Ut27dVKVKFdWvX1+jRo3SqVOnVK5cOS1atEg33HBDgfvt0KGDJk2aJIfDoe3bt+u+++5T7dq1dd1116ljx4769ttv5eXlpTfeeEPdu3f31OEVmJ+fn2655RZJ0qxZs1SpUiXde++9xTwqAMDVgjABAIBiEBERoS1btmjo0KGqW7eufH19VbFiRbVu3VqTJk1SVFSUunTp4rZt7dq1tWXLFk2bNk233XabgoODVaZMGdWqVUuPPPKItmzZogcffDBbm4kTJ2rhwoW68847FRwcLG9vbwUFBenmm2/WuHHjFBUVpddee81a0r9kyZJCOc5u3bppxowZatCggZxOp6pUqaKjR48WSt+Xq0yZMpo1a5bWr1+vBx98UGFhYSpbtqz8/PzUtGlTjRkzRlFRUQoPD7fd9wsvvKCff/5ZAwcOVJ06dVS2bFn5+vqqbt26Gjp0qDZs2KBx48Z54Kguz4IFC3T33XcrKChIvr6+SklJKe4hAQCuEg5TGNsVAwAAAACAUoOVCQAAAAAAwBbCBAAAAAAAYAthAgAAAAAAsIUwAQAAAAAA2EKYAAAAAAAAbCFMAAAAAAAAthAmAAAAAAAAWwgTAAAAAACALYQJAAAAAADAFsIEAAAAAABgC2ECAAAAAACwhTABAAAAAADYQpgAAAAAAABsIUwAAAAAAAC2ECYAAAAAAABbCBMAAAAAAIAthAkAAAAAAMAWwgQAAAAAAGALYQIAAAAAALCFMAEAAAAAANhCmAAAAAAAAGwhTAAAAAAAALYQJgAAAAAAAFsIEwAAAAAAgC3/D1xFjXH6SoC5AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1200x400 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(6,2))\n",
    "plt.hist(best_fit[\"params\"][\"local_time\"].reshape(-1).numpy(), bins=50, density=True)\n",
    "plt.xlabel(\"local time shift\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "if False:\n",
    "    plt.plot(\n",
    "        best_fit[\"mean\"][\"init_loc\"] + 0 * best_fit[\"median\"][\"init\"],\n",
    "        best_fit[\"median\"][\"init\"],\n",
    "        \"k.\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Assess model fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def plot_fits():\n",
    "    num_nonzero = int(torch.count_nonzero(weekly_strains))\n",
    "    for key, fit in fits.items():\n",
    "        median = fit.get(\"median\", fit.get(\"mean\", {}))\n",
    "        plt.figure(figsize=(8, 7))\n",
    "        time = np.arange(1, 1 + len(fit[\"losses\"]))\n",
    "        plt.plot(fit[\"losses\"], \"k--\", label=\"loss\")\n",
    "        locs = []\n",
    "        grads = []\n",
    "        for name, series in fit[\"series\"].items():\n",
    "            rankby = -torch.tensor(series).log1p().mean().item()\n",
    "            if name.startswith(\"Guide.\"):\n",
    "                name = name[len(\"Guide.\"):].replace(\"$$$\", \".\")\n",
    "                grads.append((name, series, rankby))\n",
    "            elif name.endswith(\"_centered\") or name == \"local_time\":\n",
    "                grads.append((name, series, rankby))\n",
    "            elif name != \"loss\":\n",
    "                locs.append((name, series, rankby))\n",
    "        locs.sort(key=lambda x: x[-1])\n",
    "        grads.sort(key=lambda x: x[-1])\n",
    "        for name, series, _ in locs:\n",
    "            plt.plot(time, series, label=name)\n",
    "        for name, series, _ in locs:\n",
    "            plt.plot(time, series, color=\"white\", lw=3, alpha=0.3, zorder=-1)\n",
    "        for name, series, _ in grads:\n",
    "            plt.plot(time, series, lw=1, alpha=0.3, label=name, zorder=-2)\n",
    "        plt.yscale(\"log\")\n",
    "        plt.xscale(\"log\")\n",
    "        plt.xlim(1, len(fit[\"losses\"]))\n",
    "        plt.legend(loc=\"upper left\", fontsize=8)\n",
    "        plt.xlabel(\"SVI step (duration = {:0.1f} minutes)\".format(fit[\"walltime\"]/60))\n",
    "        loss = np.median(fit[\"losses\"][-201:]) / num_nonzero\n",
    "        scalars = \" \".join([f\"L={loss:0.6g}\"] + [\n",
    "            \"{}={:0.3g}\".format(\n",
    "                \"\".join(p[0] for p in k.split(\"_\")).upper(), v\n",
    "            )\n",
    "            for k, v in median.items()\n",
    "            if v.numel() == 1\n",
    "        ])\n",
    "        plt.title(\"{} ({})\\n{}\".format(key[0], scalars, key[-1]))\n",
    "plot_fits()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Interpreting results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def plusminus(mean, std):\n",
    "    p95 = 1.96 * std\n",
    "    return torch.stack([mean - p95, mean, mean + p95])\n",
    "\n",
    "def plot_forecast(fit, queries=None, num_strains=10, filenames=[]):\n",
    "    if queries is None:\n",
    "        queries = list(location_id)\n",
    "    elif isinstance(queries, str):\n",
    "        queries = [queries]\n",
    "    fig, axes = plt.subplots(len(queries), figsize=(8, 1.5 + 2 * len(queries)), sharex=True)\n",
    "    if not isinstance(axes, (list, np.ndarray)):\n",
    "        axes = [axes]\n",
    "    dates = matplotlib.dates.date2num(mutrans.date_range(len(fit[\"mean\"][\"probs\"])))\n",
    "    forecast_steps = len(fit[\"mean\"][\"probs\"]) - len(weekly_cases)\n",
    "    assert forecast_steps >= 0\n",
    "    probs = plusminus(fit[\"mean\"][\"probs\"], fit[\"std\"][\"probs\"])  # [3, T, P, S]\n",
    "    padding = 1 + weekly_cases[-1:].expand(forecast_steps, -1)\n",
    "    weekly_cases_ = torch.cat([weekly_cases, padding], 0)\n",
    "    predicted = probs * weekly_cases_[..., None]\n",
    "    ids = torch.tensor([i for name, i in location_id.items()\n",
    "                        if any(q in name for q in queries)])\n",
    "    strain_ids = weekly_strains[:, ids].sum([0, 1]).sort(-1, descending=True).indices\n",
    "    strain_ids = strain_ids[:num_strains]\n",
    "    colors = [f\"C{i}\" for i in range(10)] + [\"black\"] * 90\n",
    "    assert len(colors) >= num_strains\n",
    "    light = \"#bbbbbb\"\n",
    "    for row, (query, ax) in enumerate(zip(queries, axes)):\n",
    "        ids = torch.tensor([i for name, i in location_id.items() if query in name])\n",
    "        print(f\"{query} matched {len(ids)} regions\")\n",
    "        counts = weekly_cases[:, ids].sum(1)\n",
    "        counts /= counts.max()\n",
    "        ax.plot(dates[:len(counts)], counts, \"k-\", color=light, lw=0.8, zorder=-20)\n",
    "        counts = weekly_strains[:, ids].sum([1, 2])\n",
    "        counts /= counts.max()\n",
    "        ax.plot(dates[:len(counts)], counts, \"k--\", color=light, lw=1, zorder=-20)\n",
    "        pred = predicted.index_select(-2, ids).sum(-2)\n",
    "        pred /= pred[1].sum(-1, True).clamp_(min=1e-8)\n",
    "        obs = weekly_strains[:, ids].sum(1)\n",
    "        obs /= obs.sum(-1, True).clamp_(min=1e-9)\n",
    "        for s, color in zip(strain_ids, colors):\n",
    "            lb, mean, ub = pred[..., s]\n",
    "            ax.fill_between(dates, lb, ub, color=color, alpha=0.2, zorder=-10)\n",
    "            ax.plot(dates, mean, color=color, lw=1, zorder=-9)\n",
    "            strain = lineage_id_inv[s]\n",
    "            ax.plot(dates[:len(obs)], obs[:, s], color=color, lw=0, marker='o', markersize=3,\n",
    "                    label=strain if row == 0 else None)\n",
    "        ax.set_ylim(0, 1)\n",
    "        ax.set_yticks(())\n",
    "        ax.set_ylabel(query.replace(\" / \", \"\\n\"))\n",
    "        ax.set_xlim(dates.min(), dates.max())\n",
    "        if row == 0:\n",
    "            ax.legend(loc=\"upper left\", fontsize=8)\n",
    "        elif row == 1:\n",
    "            ax.plot([], \"k--\", color=light, lw=1, label=\"relative #samples\")\n",
    "            ax.plot([], \"k-\", color=light, lw=0.8, label=\"relative #cases\")\n",
    "            ax.plot([], lw=0, marker='o', markersize=3, color='gray',\n",
    "                    label=\"observed portion\")\n",
    "            ax.fill_between([], [], [], color='gray', label=\"predicted portion\")\n",
    "            ax.legend(loc=\"upper left\", fontsize=8)\n",
    "    ax.xaxis.set_major_locator(matplotlib.dates.MonthLocator())\n",
    "    ax.xaxis.set_major_formatter(matplotlib.dates.DateFormatter(\"%b %Y\"))\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.subplots_adjust(hspace=0)\n",
    "    for filename in filenames:\n",
    "        plt.savefig(filename)\n",
    "\n",
    "plot_forecast(best_fit,\n",
    "              queries=[\"England\", \"USA / California\", \"Brazil\"],\n",
    "              num_strains=10,\n",
    "              filenames=[\"paper/forecast.png\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def plot_prevalence(fit, filenames=()):\n",
    "    rate = fit[\"median\"][\"rate\"]\n",
    "    if rate.dim() == 2:\n",
    "        rate = rate.mean(0)\n",
    "    rate = rate - rate[dataset[\"lineage_id\"][\"A\"]]\n",
    "    R = rate.exp()\n",
    "    probs = fit[\"mean\"][\"probs\"].mean(0)\n",
    "    cases = torch.einsum(\"ps,p->s\", probs, weekly_cases[-2])\n",
    "    cases = cases / mutrans.TIMESTEP\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    plt.scatter(R, cases, lw=0, alpha=0)\n",
    "    for name, i in lineage_id.items():\n",
    "        plt.text(R[i], cases[i], name, fontsize=8, alpha=0.8,\n",
    "                 horizontalalignment=\"center\", verticalalignment=\"center\",)\n",
    "    plt.ylabel(\"confirmed cases / day\")\n",
    "    plt.yscale(\"symlog\")\n",
    "    plt.xscale(\"log\")\n",
    "    #plt.ylim(1, None)\n",
    "    plt.xlim(0.9, None)\n",
    "    lb10 = math.floor(10 * R.min().item())\n",
    "    ub10 = math.ceil(10 * R.max().item())\n",
    "    xticks = [x10 / 10 for x10 in range(lb10, ub10 + 1)]\n",
    "    plt.xticks(xticks, list(map(str, xticks)))\n",
    "    plt.xlabel(\"relative reproduction number $R_{strain} / R_A$\")\n",
    "    plt.title(f\"Growth rate of {len(lineage_id)} PANGO lineages\")\n",
    "    plt.tight_layout()\n",
    "    for filename in filenames:\n",
    "        plt.savefig(filename)\n",
    "\n",
    "plot_prevalence(best_fit, [\"paper/strain_prevalence.png\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def plot_volcano(mean, std, filenames=(), linthresh=2, top_k=60, plot_neg=False):\n",
    "    xs = mean\n",
    "    ys = mean.abs() / std\n",
    "    assert len(xs) == len(mutations)\n",
    "    y0, y1 = float(ys.min()), float(ys.max())\n",
    "    x0, x1 = float(xs.min()), float(xs.max())\n",
    "    ys, idx = ys.sort(0, descending=True)\n",
    "    xs = xs[idx]\n",
    "    pos = (0 < xs) & (xs < math.inf)\n",
    "    neg = (-math.inf < xs) & (xs < 0)\n",
    "    ys_pos, ys_neg = ys[pos], ys[neg]\n",
    "    xs_pos, xs_neg = xs[pos], xs[neg]\n",
    "    idx_pos, idx_neg = idx[pos], idx[neg]\n",
    "    N = top_k\n",
    "\n",
    "    plt.figure(figsize=(8, 8))\n",
    "    plt.title(f\"Increased growth rate of {len(mutations)} mutations\")\n",
    "    for mask in [pos, neg]:\n",
    "        xs_mask, ys_mask = xs[mask], ys[mask]\n",
    "        plt.plot(xs_mask[:N], ys_mask[:N], 'k.', lw=0, markersize=2, zorder=10)\n",
    "        plt.plot(xs_mask[N:], ys_mask[N:], 'k.', lw=0, markersize=2, zorder=10, color=\"#aaa\")\n",
    "    plt.xlabel(\"effect size (R / R_A = exp())\")\n",
    "    xticks = [0.85, 0.9, 0.95, 1.0, 1.05, 1.1, 1.15]\n",
    "    plt.xticks(list(map(math.log, xticks)), list(map(str, xticks)))\n",
    "    plt.ylabel(\"statistical significance (||/)\")\n",
    "    if plot_neg:\n",
    "        lpad = 0.33 if any(\",\" in mutations[i] for i in idx_neg[:N].tolist()) else 0.18\n",
    "    else:\n",
    "        lpad = 0\n",
    "    rpad = 0.33 if any(\",\" in mutations[i] for i in idx_pos[:N].tolist()) else 0.18\n",
    "    plt.xlim(x0 - (x1 - x0) * lpad, x1 + (x1 - x0) * rpad)\n",
    "    plt.ylim(0, None)\n",
    "    plt.yscale(\"symlog\", linthresh=linthresh)\n",
    "    yticks = [y for y in [0, 1, 2, 5, 10, 20, 50, 100, 200, 500, 1000, 2000] if y < y1]\n",
    "    plt.yticks(yticks, list(map(str, yticks)))\n",
    "    p95 = dist.Normal(0, 1).icdf(torch.tensor(0.95)).item()\n",
    "    plt.plot([x0, x1], [p95, p95], 'k--', alpha=0.2)\n",
    "    comment_pos = 0.8 if plot_neg else 0.3\n",
    "    plt.text((1 - comment_pos) * x0 + comment_pos * x1, p95 * 0.95,\n",
    "             \"95% probabililty\\nof correct sign()\",\n",
    "             fontsize=7, horizontalalignment=\"center\", verticalalignment=\"top\",\n",
    "             alpha=0.8, zorder=100)\n",
    "        \n",
    "    colors = {\"N\": \"blue\", \"S\": \"red\", \"M\": \"purple\", \"ORF3a\": \"darkgreen\"}\n",
    "    ax = plt.gca()\n",
    "    t = (ax.transScale + ax.transLimits).inverted()\n",
    "    if plot_neg:\n",
    "        for i in range(min(N, len(xs_neg))):\n",
    "            x = x0\n",
    "            _, y = t.transform((0, 1 - (i + 1) / (N + 1)))\n",
    "            plt.plot([x, xs_neg[i]], [y, ys_neg[i]], color='gray', lw=0.2)\n",
    "            name = mutations[int(idx_neg[i])]\n",
    "            plt.text(x, y, name + \" \", color=colors.get(name.split(\":\")[0], \"gray\"),\n",
    "                     fontsize=8, verticalalignment=\"center\", horizontalalignment=\"right\")\n",
    "    for i in range(N):\n",
    "        x = x1\n",
    "        _, y = t.transform((0, 1 - (i + 1) / (N + 1)))\n",
    "        name = mutations[int(idx_pos[i])]\n",
    "        plt.plot([x, xs_pos[i]], [y, ys_pos[i]], color='gray', lw=0.2)\n",
    "        plt.text(x, y, \" \" + name, color=colors.get(name.split(\":\")[0], \"gray\"),\n",
    "                 fontsize=8, verticalalignment=\"center\", horizontalalignment=\"left\")\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_linewidth(0.5)\n",
    "    plt.tight_layout()\n",
    "    for f in filenames:\n",
    "        plt.savefig(f)\n",
    "\n",
    "plot_volcano(best_fit[\"mean\"][\"coef\"], best_fit[\"std\"][\"coef\"], linthresh=2,\n",
    "             filenames=[\"paper/volcano.png\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyrocov.sarscov2 import GENE_TO_POSITION, aa_mutation_to_position\n",
    "\n",
    "def plot_manhattan(mean, top_k=60, filenames=()):\n",
    "    assert len(mean) == len(mutations)\n",
    "    Y = mean\n",
    "    y1 = Y.max().item()\n",
    "    position = torch.tensor([aa_mutation_to_position(m) for m in mutations])\n",
    "    radius = 2 + 30 * (Y / Y.max())\n",
    "    gene_id = {gene_name: i for i, gene_name in enumerate(GENE_TO_POSITION)}\n",
    "    gene_ids = torch.tensor([gene_id[m.split(\":\")[0]] for m in mutations])\n",
    "    even = (gene_ids % 2 == 0) & (mean > 0)\n",
    "    odd = (gene_ids % 2 == 1) & (mean > 0)\n",
    "\n",
    "    plt.figure(figsize=(10, 4))\n",
    "    plt.title(f\"Increased growth rate of {len(mutations)} mutations\")\n",
    "    for mask, color in zip([even, odd], [\"darkblue\", \"darkred\"]):\n",
    "        plt.scatter(position[mask].numpy(), Y[mask].numpy(),\n",
    "                    radius[mask].numpy(), color=color, alpha=0.5, lw=0)\n",
    "    special = {\"S\": [], \"N\": []}  # Many hits, plot with lines\n",
    "    for i in Y.sort(0, descending=True).indices[:top_k].tolist():\n",
    "        x = float(position[i])\n",
    "        y = float(Y[i])\n",
    "        gene, name = mutations[i].split(\":\")\n",
    "        if gene in special:\n",
    "            special[gene].append((y, x, name))\n",
    "            continue\n",
    "        plt.text(x, y + y1/80, name, fontsize=6,\n",
    "                 verticalalignment=\"bottom\", horizontalalignment=\"center\")\n",
    "    for special_ in special.values():\n",
    "        special_.sort(reverse=True)\n",
    "    y_bounds = {k: (min(y for (y, _, _) in v), max(y for (y, _, _) in v))\n",
    "                for k, v in special.items() if v}\n",
    "    for i, (y, x, name) in enumerate(special[\"S\"]):\n",
    "        lb, ub = y_bounds[\"S\"]\n",
    "        lb, ub = lb * 0.8, ub * 0.1 + y1 * 0.9\n",
    "        y_label = 0.2 * y + 0.8 * (ub + (lb - ub) * (i / (len(special[\"S\"]) - 0.99)))\n",
    "        x_label = GENE_TO_POSITION[\"S\"][1] - 1000\n",
    "        plt.text(x_label, y_label, name, fontsize=6,\n",
    "                 verticalalignment=\"center\", horizontalalignment=\"left\")\n",
    "        plt.plot([x, x_label], [y, y_label], 'k-', lw=0.2, alpha=0.5)\n",
    "    for gene in set(special) - {\"S\"}:\n",
    "        for i, (y, x, name) in enumerate(special[gene]):\n",
    "            lb, ub = y_bounds[gene]\n",
    "            lb, ub = lb * 0.8, ub * 0.1 + y1 * 0.9\n",
    "            y_label = 0.3 * y + 0.7 * (ub + (lb - ub) * (i / (len(special[gene]) - 0.99)))\n",
    "            x_label = GENE_TO_POSITION[gene][1] + 200\n",
    "            plt.text(x_label, y_label, name, fontsize=6,\n",
    "                     verticalalignment=\"center\", horizontalalignment=\"left\")\n",
    "            plt.plot([x, x_label], [y, y_label], 'k-', lw=0.2, alpha=0.5)\n",
    "        \n",
    "    start_end = list(GENE_TO_POSITION.values())\n",
    "    plt.xlim(start_end[0][0], start_end[-1][-1])\n",
    "    xticks = []\n",
    "    for i, (gene, (start, end)) in enumerate(GENE_TO_POSITION.items()):\n",
    "        if gene == \"ORF14\":\n",
    "            continue  # skip overlapping frame\n",
    "        plt.axvline(start, lw=0.1, color=\"gray\")\n",
    "        plt.axvline(end, lw=0.1, color=\"gray\")\n",
    "        xticks.extend([start, end])\n",
    "        plt.text((start + end) / 2, -y1 / 50, gene, rotation=-90,\n",
    "                 fontsize=6, verticalalignment=\"top\", horizontalalignment=\"center\")\n",
    "    plt.xticks(xticks, labels=())\n",
    "    plt.ylim(0, None)\n",
    "    ax = plt.gca()\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_linewidth(0.1)\n",
    "    plt.ylabel(\" log R\")\n",
    "    for f in filenames:\n",
    "        plt.savefig(f)\n",
    "\n",
    "plot_manhattan(best_fit[\"mean\"][\"coef\"], top_k=55, filenames=[\"paper/manhattan.png\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_manhattan_gene(gene, mean, top_k=200, filenames=()):\n",
    "    assert len(mean) == len(mutations)\n",
    "    Y = mean\n",
    "    position = torch.tensor([aa_mutation_to_position(m) for m in mutations])\n",
    "    gene_id = {gene_name: i for i, gene_name in enumerate(GENE_TO_POSITION)}\n",
    "    gene_ids = torch.tensor([gene_id[m.split(\":\")[0]] for m in mutations])\n",
    "    mask = (gene_ids == gene_id[gene]) & (mean > 0)\n",
    "    y1 = Y[mask].max().item()\n",
    "    radius = 2 + 30 * (Y / Y[mask].max())\n",
    "\n",
    "    plt.figure(figsize=(10, 4))\n",
    "    plt.title(f\"Increased growth rate of mutations within {gene} gene\")\n",
    "    plt.scatter(position[mask].numpy(), Y[mask].numpy(),\n",
    "                radius[mask].numpy(), color=\"darkblue\", alpha=0.5, lw=0)\n",
    "    special = []  # Many hits, plot with lines\n",
    "    start, end = GENE_TO_POSITION[gene]\n",
    "    z0, z1 = {\n",
    "        # \"ORF1a\": (3500, 3750),\n",
    "        \"N\": (180, 240),\n",
    "        \"S\": (120, 180),\n",
    "    }.get(gene, (0, 0))\n",
    "    for i in Y.sort(0, descending=True).indices[:top_k].tolist():\n",
    "        x = float(position[i])\n",
    "        y = float(Y[i])\n",
    "        g, name = mutations[i].split(\":\")\n",
    "        pos = int(re.search(\"[0-9]+\", name).group())\n",
    "        if g != gene:\n",
    "            continue\n",
    "        if z0 < pos < z1:\n",
    "            special.append((y, x, name))\n",
    "        else:\n",
    "            plt.text(x, y * 1.01 + + y1 / 60, name, fontsize=6,\n",
    "                     verticalalignment=\"center\", horizontalalignment=\"center\")\n",
    "    special.sort(reverse=True)\n",
    "    if special:\n",
    "        lb = min(y for (y, _, _) in special)\n",
    "        ub = max(y for (y, _, _) in special)\n",
    "        lb, ub = lb * 0.5, ub * 0.1 + y1 * 0.9\n",
    "        for i, (y, x, name) in enumerate(special):\n",
    "            y_label = 0.1 * y + 0.9 * (ub + (lb - ub) * (i / (len(special) - 0.99)))\n",
    "            x_label = start + 3 * z1\n",
    "            plt.text(x_label, y_label, name, fontsize=6,\n",
    "                     verticalalignment=\"center\", horizontalalignment=\"left\")\n",
    "            plt.plot([x, x_label], [y, y_label], 'k-', lw=0.2, alpha=0.5)\n",
    "    xticks = [start]\n",
    "    dx = 150\n",
    "    while end - start > dx * 12:\n",
    "        dx *= 2\n",
    "    while xticks[-1] + dx < end:\n",
    "        xticks.append(xticks[-1] + dx)\n",
    "    labels = [str((x - start) // 3) for x in xticks]\n",
    "    plt.xticks(xticks, labels)\n",
    "    plt.xlabel(f\"amino acid position within {gene} gene\")\n",
    "    plt.xlim(start, end)\n",
    "    plt.ylim(0, None)\n",
    "    ax = plt.gca()\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_linewidth(0.1)\n",
    "    plt.ylabel(\" log R\")\n",
    "    for f in filenames:\n",
    "        plt.savefig(f)\n",
    "\n",
    "plot_manhattan_gene(\"ORF1a\", best_fit[\"mean\"][\"coef\"], 100, filenames=[\"paper/manhattan_ORF1a.png\"])\n",
    "plot_manhattan_gene(\"S\", best_fit[\"mean\"][\"coef\"], 300, filenames=[\"paper/manhattan_S.png\"])\n",
    "plot_manhattan_gene(\"N\", best_fit[\"mean\"][\"coef\"], 350, filenames=[\"paper/manhattan_N.png\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extracting a table of top mutations and their stories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rank_mutations(fit, names):\n",
    "    mean = fit[\"mean\"][\"coef\"]\n",
    "    std = fit[\"std\"][\"coef\"]\n",
    "    sigma = mean / std.clamp(min=1e-8)\n",
    "    ranks = sigma.sort(0, descending=True).indices.tolist()\n",
    "    assert len(ranks) == len(mutations)\n",
    "    ranked = [mutations[k] for k in ranks]\n",
    "    ranks = {m: i for i, m in enumerate(ranked)}\n",
    "    print(\"Mut'n\\tRank\\tEstimate\")\n",
    "    for name in names:\n",
    "        i = mutations.index(name)\n",
    "        print(\"{}\\t{}\\t{:0.3g}  {:0.2g}\".format(name, ranks[name], mean[i], std[i]))\n",
    "\n",
    "rank_mutations(best_fit, [\"S:D614G\", \"S:N501Y\", \"S:E484K\", \"S:L452R\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parent_features = torch.zeros_like(features)\n",
    "for c, child in enumerate(lineage_id_inv):\n",
    "    child = pangolin.decompress(child)\n",
    "    parent = child\n",
    "    while True:\n",
    "        parent = \"A\" if parent == \"A\" else pangolin.get_parent(parent)\n",
    "        try:\n",
    "            p = lineage_id[pangolin.compress(parent)]\n",
    "            break\n",
    "        except KeyError:\n",
    "            continue\n",
    "    parent_features[c] = features[p]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_emergences(i):\n",
    "    delta = features[:, i] - parent_features[:, i]\n",
    "    emerged = set((delta > 0.5).nonzero(as_tuple=True)[0].tolist())\n",
    "    emerged.add(delta.argmax().item())\n",
    "    result = []\n",
    "    for k in sorted(emerged):\n",
    "        name = lineage_id_inv[k]\n",
    "        longname = pangolin.decompress(name)\n",
    "        result.append(name if name == longname else f\"{name} ({longname})\")\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_mutation_table(fit, filename):\n",
    "    assert filename.endswith(\".tsv\")\n",
    "    mean = fit[\"mean\"][\"coef\"]\n",
    "    std = fit[\"std\"][\"coef\"]\n",
    "    sigma = mean / std.clamp(min=1e-8)\n",
    "    if \"samples\" in fit:\n",
    "        lb, ub = stats.confidence_interval(0.95, fit[\"samples\"][\"coef\"])\n",
    "    else:\n",
    "        lb, ub = dist.Normal(mean, std).icdf(torch.tensor([0.025, 0.975])[:, None])\n",
    "    R_RA = mean.exp()  # mean is in units of generation time.\n",
    "    lineage_counts = weekly_strains.sum((0, 1))\n",
    "    schema = [\n",
    "        (\"rank\", \"{:d}\"),\n",
    "        (\"mutation\", \"{:s}\"),\n",
    "        (\"mean/stddev\", \"{:0.6g}\"),\n",
    "        (\" log R\", \"{:0.6g}\"),\n",
    "        (\" log R 95% ci lower\", \"{:0.6g}\"),\n",
    "        (\" log R 95% ci upper\", \"{:0.6g}\"),\n",
    "        (\"R / R_A\", \"{:0.6g}\"),\n",
    "        (\"R / R_A 95% ci lower\", \"{:0.6g}\"),\n",
    "        (\"R / R_A 95% ci lower\", \"{:0.6g}\"),\n",
    "        (\"emerged in lineages\", \"{:s}\"),  \n",
    "    ]\n",
    "    header = \"\\t\".join(h for h, r in schema) + \"\\n\"\n",
    "    row = \"\\t\".join(r for h, r in schema) + \"\\n\"\n",
    "    with open(filename, \"wt\") as f:\n",
    "        f.write(header)\n",
    "        for rank, i in enumerate(sigma.sort(0, descending=True).indices.tolist()):\n",
    "            emerged = find_emergences(i)\n",
    "            f.write(row.format(\n",
    "                rank + 1, mutations[i],\n",
    "                sigma[i], mean[i], lb[i], ub[i], R_RA[i], lb[i].exp(), ub[i].exp(),\n",
    "                \", \".join(emerged)\n",
    "            ))\n",
    "\n",
    "write_mutation_table(best_fit, \"paper/mutations.tsv\")\n",
    "pd.read_csv(\"paper/mutations.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(columns.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimate_lineage_bdays(exclude_first=10):\n",
    "    lineage_days = defaultdict(list)\n",
    "    for lineage, day in zip(columns[\"lineage\"], columns[\"day\"]):\n",
    "        lineage = pangolin.decompress(lineage)\n",
    "        lineage_days[lineage].append(day)\n",
    "    lineage_bday = {}\n",
    "    for lineage, days in list(lineage_days.items()):\n",
    "        days.sort()\n",
    "        lineage_bday[lineage] = days[min(exclude_first, len(days) // 10)]\n",
    "    start_date = datetime.datetime.strptime(mutrans.START_DATE, \"%Y-%m-%d\")\n",
    "    return {\n",
    "        lineage: (start_date + datetime.timedelta(days=day)).strftime(\"%Y-%m-%d\")\n",
    "        for lineage, day in lineage_bday.items()\n",
    "    }\n",
    "\n",
    "lineage_bday = estimate_lineage_bdays()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_strain_table(fit, filename):\n",
    "    assert filename.endswith(\".tsv\")\n",
    "    rate_mean = fit[\"mean\"][\"rate\"].mean(0)\n",
    "    rate_std = (fit[\"std\"][\"rate\"].mean(0) ** 2 + fit[\"mean\"][\"rate\"].std(0) ** 2).sqrt()\n",
    "    sigma = rate_mean / rate_std.clamp(min=1e-8)\n",
    "    R_mean = rate_mean.exp()\n",
    "    lb, ub = dist.Normal(rate_mean, rate_std).icdf(torch.tensor([0.025, 0.975])[..., None])\n",
    "    RA = R_mean[lineage_id[\"A\"]].item()\n",
    "    R_RA = R_mean / RA\n",
    "    lb = lb / RA\n",
    "    ub = ub / RA\n",
    "    probs = fit[\"median\"][\"probs\"][:len(weekly_cases)]\n",
    "    cases = torch.einsum(\"tps,tp->ts\", probs, weekly_cases)\n",
    "    cases_per_day = cases[-2] / mutrans.TIMESTEP\n",
    "    cases_total = cases.sum(0)\n",
    "    schema = [\n",
    "        (\"rank\", \"{:d}\"),\n",
    "        (\"strain\", \"{:s}\"),\n",
    "        (\"mean/stddev\", \"{:.6g}\"),\n",
    "        (\"R / R_A\", \"{:.6g}\"),\n",
    "        (\"R / R_A 95% ci lower\", \"{:.6g}\"),\n",
    "        (\"R / R_A 95% ci upper\", \"{:.6g}\"),\n",
    "        (\"confirmed cases / day\", \"{:.6g}\"),\n",
    "        (\"confirmed cases total\", \"{:.6g}\"),\n",
    "        (\"birthday\", \"{:s}\"),\n",
    "        (\"mutations\", \"{:s}\"),\n",
    "    ]\n",
    "    header = \"\\t\".join(h for h, r in schema) + \"\\n\"\n",
    "    row = \"\\t\".join(r for h, r in schema) + \"\\n\"\n",
    "    with open(filename, \"wt\") as f:\n",
    "        f.write(header)\n",
    "        for rank, i in enumerate(R_RA.sort(0, descending=True).indices.tolist()):\n",
    "            lineage = lineage_id_inv[i]\n",
    "            lineage_mutations = \",\".join(\n",
    "                mutations[m] for m in features[i].nonzero(as_tuple=True)[0].tolist()\n",
    "            )\n",
    "            f.write(row.format(\n",
    "                rank + 1, lineage,\n",
    "                sigma[i], R_RA[i], lb[i], ub[i], cases_per_day[i], cases_total[i],\n",
    "                lineage_bday[pangolin.decompress(lineage)],\n",
    "                lineage_mutations,\n",
    "            ))\n",
    "\n",
    "write_strain_table(best_fit, \"paper/strains.tsv\")\n",
    "pd.read_csv(\"paper/strains.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison with deep mutational scans"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first compare with [(Starr et al. 2020)](https://www.sciencedirect.com/science/article/pii/S0092867420310035) who study S mutations affecting folding and ACE2 binding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"data/mutation-studies/1-s2.0-S0092867420310035-mmc2.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "folding = {f\"S:{m}\": float(e) for m, e in zip(df[\"mutation\"], df[\"expr_avg\"])}\n",
    "binding = {f\"S:{m}\": float(b) for m, b in zip(df[\"mutation\"], df[\"bind_avg\"])}\n",
    "print(sum(1 for m in mutations if m in folding))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We next compare with [(Greaney et al. 2021)](https://www.sciencedirect.com/science/article/pii/S1931312820306247) who study antibody escape."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"data/mutation-studies/1-s2.0-S1931312820306247-mmc2.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "escape = {\n",
    "    f\"S:{w}{s}{m}\": float(e)\n",
    "    for w, s, m, e in zip(df[\"wildtype\"], df[\"site\"], df[\"mutation\"], df[\"mut_escape\"])\n",
    "}\n",
    "print(sum(1 for m in mutations if m in escape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 3, figsize=(10, 3), sharey=True)\n",
    "ms = [m for m in mutations if m in escape]\n",
    "y = best_fit[\"mean\"][\"coef\"][[i for i, m in enumerate(mutations) if m in escape]].numpy()\n",
    "axes[0].set_ylabel(\" log R\")\n",
    "for name, ax in zip([\"folding\", \"binding\", \"escape\"], axes):\n",
    "    scan = locals()[name]\n",
    "    x = torch.tensor([scan[m] for m in ms]).numpy()\n",
    "    # ax.scatter(x, y, alpha=0.5, lw=0)\n",
    "    for xm, ym, m in zip(x, y, ms):\n",
    "        ax.text(xm, ym, m[2:], fontsize=6,\n",
    "                verticalalignment=\"center\", horizontalalignment=\"center\")\n",
    "    ax.set_xlim(1.08 * x.min() - 0.08 * x.max(), 1.08 * x.max() - 0.08 * x.min())\n",
    "    ax.set_ylim(1.05 * y.min() - 0.05 * y.max(), 1.05 * y.max() - 0.05 * y.min())\n",
    "    ax.set_xlabel(f\"{name} ( = {pearson_correlation(x, y):0.2g})\")\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_linewidth(0.5)\n",
    "axes[2].set_xscale(\"log\")\n",
    "axes[2].set_xlim(x.min() ** 1.08 / x.max() ** 0.08, x.max() ** 1.08 / x.min() ** 0.08)\n",
    "axes[1].set_title(f\"Comparison of {len(ms)} S gene mutations to deep scanning\")\n",
    "plt.subplots_adjust(wspace=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It's hard to say whether these correlations are meaningful, as they are dominated by a few outliers.\n",
    "\n",
    "Let's fit a linear model regressing growth rate against theses deep scanning results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyro\n",
    "import pyro.distributions as dist\n",
    "from pyro.infer import SVI, Trace_ELBO\n",
    "from pyro.infer.autoguide import AutoMultivariateNormal\n",
    "from pyro.optim import Adam\n",
    "\n",
    "def fit_model(fit):\n",
    "    trans_data = fit[\"mean\"][\"coef\"][[i for i, m in enumerate(mutations) if m in escape]]\n",
    "    folding_data = torch.tensor([folding[m] for m in ms])\n",
    "    binding_data = torch.tensor([binding[m] for m in ms])\n",
    "    escape_data = torch.tensor([escape[m] for m in ms])\n",
    "    \n",
    "    def model():\n",
    "        coef = pyro.sample(\"coef\", dist.Normal(0, 10).expand([5]).to_event(1))\n",
    "        t, f, b, e, be = coef.unbind(-1)\n",
    "        noise = pyro.sample(\"noise\", dist.LogNormal(0, 2))\n",
    "        with pyro.plate(\"data\", len(trans_data)):\n",
    "            pred = (\n",
    "                t + f * folding_data + b * binding_data + e * escape_data\n",
    "                + be * binding_data * escape_data\n",
    "            )\n",
    "            pyro.sample(\"trans\", dist.Normal(pred, noise), obs=trans_data)\n",
    "\n",
    "    pyro.clear_param_store()\n",
    "    guide = AutoMultivariateNormal(model)\n",
    "    elbo = Trace_ELBO(num_particles=100, vectorize_particles=True)\n",
    "    svi = SVI(model, guide, Adam({\"lr\": 0.2}), elbo)\n",
    "    for step in range(201):\n",
    "        loss = svi.step()\n",
    "        if step % 20 == 0:\n",
    "            print(f\"step {step} loss = {loss:0.4g}\")\n",
    "    loc, scale = guide._loc_scale()\n",
    "    print(\"Model:\")\n",
    "    print(\"growth rate = t + f folding + b binding + e escape + be binding escape\")\n",
    "    print(\"Learned coefficients:\")\n",
    "    for k, l, s in zip(\"t f b e be\".split(), loc.tolist(), scale.tolist()):\n",
    "        print(f\"{k} = {l:0.4g} +- {s:0.2f}\")\n",
    "        \n",
    "fit_model(best_fit)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fit on subsets of data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Counter([n for n in columns[\"virus_name\"] if \"-CDC-2-\" in n]).most_common(2))\n",
    "print(Counter([n for n in columns[\"location\"] if \"USA\" in n]).most_common(2))\n",
    "print(Counter([n for n in columns[\"location\"] if \"United King\" in n]).most_common(2))\n",
    "print(Counter([n for n in columns[\"location\"] if \"North America \" in n]).most_common(2))\n",
    "print(Counter([n for n in columns[\"location\"] if \"Europe \" in n]).most_common(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "holdout_fits = {k[-1]: v for k, v in fits.items() if k[-1]}\n",
    "for key in holdout_fits:\n",
    "    print(key[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aliases = [\n",
    "    \"full data\",\n",
    "    \"only Europe\",\n",
    "    \"excluding Europe\",\n",
    "    # \"only North America\",\n",
    "    # \"excluding North America\",\n",
    "    # \"only the USA\",\n",
    "    # \"excluding the USA\",\n",
    "    # \"excluding the UK\",\n",
    "    # \"only the UK\",\n",
    "    # \"only CDC data\",\n",
    "    # \"only CDC NS3 data\",\n",
    "]\n",
    "holdout_fits = dict(zip(aliases, [best_fit] + list(holdout_fits.values())))\n",
    "assert len(holdout_fits) == 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_mutation_agreements(holdouts, filenames=()):\n",
    "    def get_mean(fit):\n",
    "        return fit[\"mean\"][\"coef\"]\n",
    "    (name0, fit0), (name1, fit1), (name2, fit2) = holdouts.items()\n",
    "    pairs = [\n",
    "        [(name0, fit0), (name1, fit1)],\n",
    "        [(name0, fit0), (name2, fit2)],\n",
    "        [(name1, fit1), (name2, fit2)],\n",
    "    ]\n",
    "    fig, axes = plt.subplots(1, 3, figsize=(10, 3), sharey=True)\n",
    "    means = [get_mean(fit) for fit in holdouts.values()]\n",
    "    x0 = min(mean.min().item() for mean in means)\n",
    "    x1 = max(mean.max().item() for mean in means)\n",
    "    lb = 1.05 * x0 - 0.05 * x1\n",
    "    ub = 1.05 * x1 - 0.05 * x0\n",
    "    fig.suptitle(\" log R for {} mutations estimated from subsets of data\"\n",
    "                 .format(len(list(holdouts.values())[0][\"mutations\"])))\n",
    "    for ax, ((name1, fit1), (name2, fit2)) in zip(axes, pairs):\n",
    "        mutations = sorted(set(fit1[\"mutations\"]) & set(fit2[\"mutations\"]))\n",
    "        means = []\n",
    "        for fit in (fit1, fit2):\n",
    "            m_to_i = {m: i for i, m in enumerate(fit[\"mutations\"])}\n",
    "            idx = torch.tensor([m_to_i[m] for m in mutations])\n",
    "            means.append(get_mean(fit)[idx])\n",
    "        ax.plot([lb, ub], [lb, ub], 'k--', alpha=0.3, zorder=-100)\n",
    "        ax.scatter(means[1].numpy(), means[0].numpy(), 50, alpha=1, lw=0, color=\"white\")\n",
    "        ax.scatter(means[1].numpy(), means[0].numpy(), 30, alpha=0.3, lw=0, color=\"darkred\")\n",
    "        ax.text(x0, 0.05 * x0 + 0.95 * x1,\n",
    "                \" = {:0.2g}\".format(pearson_correlation(means[0], means[1])))\n",
    "        ax.set_xlim(lb, ub)\n",
    "        ax.set_ylim(lb, ub)\n",
    "        ax.set_xlabel(name2)\n",
    "        ax.set_ylabel(name1)\n",
    "    plt.subplots_adjust(wspace=0.12)\n",
    "    for f in filenames:\n",
    "        plt.savefig(f)\n",
    "\n",
    "plot_mutation_agreements(holdout_fits, [\"paper/mutation_agreement.png\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_strain_agreements(fit1, holdouts, filenames=()):\n",
    "    def get_mean(fit):\n",
    "        rate = fit[\"median\"][\"rate\"]\n",
    "        if rate.dim() == 2:\n",
    "            rate = rate.median(0).values\n",
    "        return (rate - rate[lineage_id[\"A\"]]).exp()\n",
    "    (name0, fit0), (name1, fit1), (name2, fit2) = holdouts.items()\n",
    "    pairs = [\n",
    "        [(name0, fit0), (name1, fit1)],\n",
    "        [(name0, fit0), (name2, fit2)],\n",
    "        [(name1, fit1), (name2, fit2)],\n",
    "    ]\n",
    "    fig, axes = plt.subplots(1, 3, figsize=(10, 3), sharey=True)\n",
    "    means = [get_mean(fit) for fit in holdouts.values()]\n",
    "    x0 = min(mean.min().item() for mean in means)\n",
    "    x1 = max(mean.max().item() for mean in means)\n",
    "    lb = 1.05 * x0 - 0.05 * x1\n",
    "    ub = 1.05 * x1 - 0.05 * x0\n",
    "    fig.suptitle(\"$R_{{lineage}} / R_A$ for {} lineages esimated from subsets of data\"\n",
    "                 .format(len(lineage_id)))\n",
    "    for ax, ((name1, fit1), (name2, fit2)) in zip(axes, pairs):\n",
    "        mutations = sorted(set(fit1[\"mutations\"]) & set(fit2[\"mutations\"]))\n",
    "        means = [get_mean(fit) for fit in (fit1, fit2)]\n",
    "        ax.plot([lb, ub], [lb, ub], 'k--', alpha=0.3, zorder=-100)\n",
    "        ax.scatter(means[1].numpy(), means[0].numpy(), 50, alpha=1, lw=0, color=\"white\")\n",
    "        ax.scatter(means[1].numpy(), means[0].numpy(), 30, alpha=0.3, lw=0, color=\"darkred\")\n",
    "        ax.text(x0, 0.05 * x0 + 0.95 * x1,\n",
    "                \" = {:0.2g}\".format(pearson_correlation(means[0], means[1])))\n",
    "        ax.set_xlim(lb, ub)\n",
    "        ax.set_ylim(lb, ub)\n",
    "        ax.set_xlabel(name1)\n",
    "        ax.set_ylabel(name2)\n",
    "    plt.subplots_adjust(wspace=0.12)\n",
    "    for f in filenames:\n",
    "        plt.savefig(f)\n",
    "\n",
    "plot_strain_agreements(best_fit, holdout_fits, [\"paper/lineage_agreement.png\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_mutation_subset_boxplot(fits, rankby=\"t\", top_k=20, filenames=()):\n",
    "    best_fit = list(fits.values())[1]\n",
    "    if rankby == \"s\":\n",
    "        rankby = best_fit[\"mean\"][\"coef\"] / best_fit[\"std\"][\"coef\"]\n",
    "        title = f\"Top {top_k} most statistically significant mutations\"\n",
    "    elif rankby == \"t\":\n",
    "        rankby = best_fit[\"mean\"][\"coef\"]\n",
    "        title = f\"Top {top_k} mutations associated with increased growth rate\"\n",
    "    else: raise ValueError(rankby)\n",
    "    top_indices = rankby.sort(0, descending=True).indices[:top_k]\n",
    "    top_mutations = [mutations[i] for i in top_indices.tolist()]\n",
    "    xscale = 0.6\n",
    "    positions = (torch.arange(top_k)[:, None] * (len(fits) + 1)\n",
    "                 + torch.arange(len(fits))).reshape(-1) * xscale\n",
    "    data = [None] * top_k * len(fits)\n",
    "    lines = [None] * top_k * (len(fits) + 1)\n",
    "    for j, fit in enumerate(fits.values()):\n",
    "        if \"samples\" in fit:\n",
    "            samples = fit[\"samples\"][\"coef\"].T[top_indices].T\n",
    "        else:\n",
    "            mean = fit[\"mean\"][\"coef\"][top_indices]\n",
    "            std = fit[\"std\"][\"coef\"][top_indices]\n",
    "            samples = dist.Normal(mean, std).sample((1000,))\n",
    "        for i in range(top_k):\n",
    "            data[i * len(fits) + j] = samples[:, i]\n",
    "            lines[i * (len(fits) + 1) + j] = samples[:, i].mean(0)\n",
    "    xs = [None if y is None else i * xscale for i, y in enumerate(lines)]\n",
    "    \n",
    "    plt.figure(figsize=(8, 3))\n",
    "    plt.plot(xs, lines, \"k-\", alpha=0.2)\n",
    "    props = {\"linewidth\": 0.5}\n",
    "    boxplot = plt.boxplot(data, positions=positions, vert=True, patch_artist=True,\n",
    "                          showfliers=False,\n",
    "                          boxprops=props, whiskerprops=props, capprops=props,\n",
    "                          medianprops={\"alpha\": 0})\n",
    "    colors = ['lightblue', 'pink', 'lightgreen']\n",
    "    for i, patch in enumerate(boxplot['boxes']):\n",
    "        patch.set_facecolor(colors[i % len(fits)])\n",
    "    for name, c in zip(fits, colors):\n",
    "        plt.plot([], label=name, marker=\"s\", color=c)\n",
    "    plt.legend(loc=\"best\", prop={'size': 9})\n",
    "    start = (len(fits) - 1) / 2\n",
    "    plt.xticks(torch.linspace(start, start + (top_k - 1) * (len(fits) + 1), top_k) * xscale,\n",
    "               # labels=[x.replace(\":\", \":\\n\") for x in top_mutations],\n",
    "               labels=top_mutations, rotation=-90, fontsize=9)\n",
    "    plt.ylabel(\" log R\")\n",
    "    plt.title(title)\n",
    "    for f in filenames:\n",
    "        plt.savefig(f)\n",
    "\n",
    "for name in [\"Europe\"]:\n",
    "    name_ = name.lower().replace(\" \", \"_\")\n",
    "    plot_mutation_subset_boxplot({\n",
    "        f\"World w/o {name}\": holdout_fits[f\"excluding {name}\"],\n",
    "        \"World\": best_fit,\n",
    "        f\"{name} only\": holdout_fits[f\"only {name}\"],\n",
    "    }, filenames=[f\"paper/mutation_{name_}_boxplot.png\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_strain_subset_boxplot(fits, top_k=20, filenames=()):\n",
    "    def get_rate(stats):\n",
    "        rate = stats[\"rate\"]\n",
    "        if rate.dim() == 2:\n",
    "            rate = rate.median(0).values\n",
    "        return rate\n",
    "    best_fit = list(fits.values())[1]\n",
    "    top_indices = get_rate(best_fit[\"median\"]).sort(0, descending=True).indices[:top_k]\n",
    "    top_lineages = [lineage_id_inv[i] for i in top_indices.tolist()]\n",
    "    xscale = 0.6\n",
    "    positions = (torch.arange(top_k)[:, None] * (len(fits) + 1)\n",
    "                 + torch.arange(len(fits))).reshape(-1) * xscale\n",
    "    data = [None] * top_k * len(fits)\n",
    "    lines = [None] * top_k * (len(fits) + 1)\n",
    "    for j, fit in enumerate(fits.values()):\n",
    "        mean = get_rate(fit[\"median\"])[top_indices]\n",
    "        std = get_rate(fit[\"std\"])[top_indices]\n",
    "        samples = dist.Normal(mean, std).sample((1000,))\n",
    "        samples = samples - get_rate(fit[\"median\"])[lineage_id[\"A\"]]\n",
    "        samples = samples.exp()\n",
    "        for i in range(top_k):\n",
    "            data[i * len(fits) + j] = samples[:, i]\n",
    "            lines[i * (len(fits) + 1) + j] = samples[:, i].mean()\n",
    "    xs = [None if y is None else i * xscale for i, y in enumerate(lines)]\n",
    "\n",
    "    plt.figure(figsize=(8, 3))\n",
    "    plt.plot(xs, lines, \"k-\", alpha=0.2)\n",
    "    props = {\"linewidth\": 0.5}\n",
    "    boxplot = plt.boxplot(data, positions=positions, vert=True, patch_artist=True,\n",
    "                          showfliers=False,\n",
    "                          boxprops=props, whiskerprops=props, capprops=props,\n",
    "                          medianprops={\"alpha\": 0})\n",
    "    colors = ['lightblue', 'pink', 'lightgreen']\n",
    "    for i, patch in enumerate(boxplot['boxes']):\n",
    "        patch.set_facecolor(colors[i % len(fits)])\n",
    "    for name, c in zip(fits, colors):\n",
    "        plt.plot([], label=name, marker=\"s\", color=c)\n",
    "    plt.legend(loc=\"best\", prop={'size': 10})\n",
    "    start = (len(fits) - 1) / 2\n",
    "    plt.xticks(torch.linspace(start, start + (top_k - 1) * (len(fits) + 1), top_k) * xscale,\n",
    "               labels=top_lineages, rotation=-90)\n",
    "    plt.ylabel(\"$R / R_A$\")\n",
    "    plt.title(f\"Top {top_k} lineages with highest growth rate\")\n",
    "    for f in filenames:\n",
    "        plt.savefig(f)\n",
    "\n",
    "for name in [\"Europe\"]:\n",
    "    name_ = name.lower().replace(\" \", \"_\")\n",
    "    plot_strain_subset_boxplot({\n",
    "        f\"World w/o {name}\": holdout_fits[f\"excluding {name}\"],\n",
    "        \"World\": best_fit,\n",
    "        f\"{name} only\": holdout_fits[f\"only {name}\"],\n",
    "    }, filenames=[f\"paper/strain_{name_}_boxplot.png\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
